--- buildroot/boot/at91bootstrap/at91bootstrap.mk
+++ buildroot/boot/at91bootstrap/at91bootstrap.mk
@@ -26,6 +26,15 @@
 	rmdir $(BUILD_DIR)/Bootstrap-v$(AT91BOOTSTRAP_VERSION)
 endef
 
+ifneq ($(call qstrip,$(BR2_TARGET_AT91BOOTSTRAP_CUSTOM_PATCH_DIR)),)
+define AT91BOOTSTRAP_APPLY_CUSTOM_PATCHES
+	support/scripts/apply-patches.sh $(@D) $(BR2_TARGET_AT91BOOTSTRAP_CUSTOM_PATCH_DIR) \
+		at91bootstrap-$(AT91BOOTSTRAP_VERSION)-\*.patch
+endef
+
+AT91BOOTSTRAP_POST_PATCH_HOOKS += AT91BOOTSTRAP_APPLY_CUSTOM_PATCHES
+endif
+
 define AT91BOOTSTRAP_BUILD_CMDS
 	$(MAKE) CROSS_COMPILE=$(TARGET_CROSS) -C $(@D)/$(AT91BOOTSTRAP_MAKE_SUBDIR)
 endef
--- buildroot/boot/at91bootstrap/Config.in
+++ buildroot/boot/at91bootstrap/Config.in
@@ -9,6 +9,15 @@
 	  - Physical media algorithm such as DataFlash, NandFlash, NOR Flash...
 
 if	BR2_TARGET_AT91BOOTSTRAP
+
+config BR2_TARGET_AT91BOOTSTRAP_CUSTOM_PATCH_DIR
+	string "custom patch dir"
+	help
+	  If your board requires custom patches, add the path to the
+	  directory containing the patches here. The patches must be
+	  named at91bootstrap-<version>-<something>.patch.
+
+	  Most users may leave this empty
 
 config BR2_TARGET_AT91BOOTSTRAP_BOARD
 	string "Bootstrap board"
