diff -Naur old/package/bluez_utils/bluez_utils.mk new/package/bluez_utils/bluez_utils.mk
--- old/package/bluez_utils/bluez_utils.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/bluez_utils/bluez_utils.mk	2012-09-03 17:14:40.875000000 +0100
@@ -3,12 +3,14 @@
 # bluez_utils
 #
 #############################################################
-BLUEZ_UTILS_VERSION = 4.96
+
+BLUEZ_UTILS_VERSION = 4.101
 BLUEZ_UTILS_SOURCE = bluez-$(BLUEZ_UTILS_VERSION).tar.gz
-BLUEZ_UTILS_SITE = http://www.kernel.org/pub/linux/bluetooth/
+BLUEZ_UTILS_SITE = $(BR2_KERNEL_MIRROR)/linux/bluetooth
 BLUEZ_UTILS_INSTALL_STAGING = YES
 BLUEZ_UTILS_DEPENDENCIES = dbus libglib2
 BLUEZ_UTILS_CONF_OPT = --enable-test --enable-tools
+BLUEZ_UTILS_AUTORECONF = YES
 
 # BlueZ 3.x compatibility
 ifeq ($(BR2_PACKAGE_BLUEZ_UTILS_COMPAT),y)
diff -Naur old/package/busybox/busybox-1.19.x.config new/package/busybox/busybox-1.19.x.config
--- old/package/busybox/busybox-1.19.x.config	2011-11-30 23:08:44.000000000 +0000
+++ new/package/busybox/busybox-1.19.x.config	2012-10-22 21:26:01.156250000 +0100
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
-# Busybox version: 1.19.0
-# Tue Aug 23 11:28:35 2011
+# Busybox version: 1.19.3
+# Mon Oct 22 16:20:20 2012
 #
 CONFIG_HAVE_DOT_CONFIG=y
 
@@ -14,7 +14,7 @@
 #
 CONFIG_DESKTOP=y
 # CONFIG_EXTRA_COMPAT is not set
-CONFIG_INCLUDE_SUSv2=y
+# CONFIG_INCLUDE_SUSv2 is not set
 # CONFIG_USE_PORTABLE_CODE is not set
 CONFIG_PLATFORM_LINUX=y
 CONFIG_FEATURE_BUFFERS_USE_MALLOC=y
@@ -22,7 +22,7 @@
 # CONFIG_FEATURE_BUFFERS_GO_IN_BSS is not set
 CONFIG_SHOW_USAGE=y
 CONFIG_FEATURE_VERBOSE_USAGE=y
-# CONFIG_FEATURE_COMPRESS_USAGE is not set
+CONFIG_FEATURE_COMPRESS_USAGE=y
 CONFIG_FEATURE_INSTALLER=y
 # CONFIG_INSTALL_NO_USR is not set
 # CONFIG_LOCALE_SUPPORT is not set
@@ -49,7 +49,7 @@
 # CONFIG_FEATURE_PREFER_APPLETS is not set
 CONFIG_BUSYBOX_EXEC_PATH="/proc/self/exe"
 CONFIG_FEATURE_SYSLOG=y
-# CONFIG_FEATURE_HAVE_RPC is not set
+CONFIG_FEATURE_HAVE_RPC=y
 
 #
 # Build Options
@@ -60,7 +60,7 @@
 # CONFIG_BUILD_LIBBUSYBOX is not set
 # CONFIG_FEATURE_INDIVIDUAL is not set
 # CONFIG_FEATURE_SHARED_BUSYBOX is not set
-# CONFIG_LFS is not set
+CONFIG_LFS=y
 CONFIG_CROSS_COMPILER_PREFIX=""
 CONFIG_EXTRA_CFLAGS=""
 
@@ -93,7 +93,7 @@
 CONFIG_FEATURE_RTMINMAX=y
 CONFIG_PASSWORD_MINLEN=6
 CONFIG_MD5_SIZE_VS_SPEED=2
-# CONFIG_FEATURE_FAST_TOP is not set
+CONFIG_FEATURE_FAST_TOP=y
 # CONFIG_FEATURE_ETC_NETWORKS is not set
 CONFIG_FEATURE_USE_TERMIOS=y
 CONFIG_FEATURE_EDITING=y
@@ -105,8 +105,8 @@
 CONFIG_FEATURE_TAB_COMPLETION=y
 # CONFIG_FEATURE_USERNAME_COMPLETION is not set
 CONFIG_FEATURE_EDITING_FANCY_PROMPT=y
-# CONFIG_FEATURE_EDITING_ASK_TERMINAL is not set
-CONFIG_FEATURE_NON_POSIX_CP=y
+CONFIG_FEATURE_EDITING_ASK_TERMINAL=y
+# CONFIG_FEATURE_NON_POSIX_CP is not set
 # CONFIG_FEATURE_VERBOSE_CP_MESSAGE is not set
 CONFIG_FEATURE_COPYBUF_KB=4
 CONFIG_FEATURE_SKIP_ROOTFS=y
@@ -123,14 +123,14 @@
 #
 # CONFIG_FEATURE_SEAMLESS_XZ is not set
 # CONFIG_FEATURE_SEAMLESS_LZMA is not set
-# CONFIG_FEATURE_SEAMLESS_BZ2 is not set
-# CONFIG_FEATURE_SEAMLESS_GZ is not set
+CONFIG_FEATURE_SEAMLESS_BZ2=y
+CONFIG_FEATURE_SEAMLESS_GZ=y
 # CONFIG_FEATURE_SEAMLESS_Z is not set
 CONFIG_AR=y
 # CONFIG_FEATURE_AR_LONG_FILENAMES is not set
 CONFIG_FEATURE_AR_CREATE=y
 CONFIG_BUNZIP2=y
-# CONFIG_BZIP2 is not set
+CONFIG_BZIP2=y
 CONFIG_CPIO=y
 # CONFIG_FEATURE_CPIO_O is not set
 # CONFIG_FEATURE_CPIO_P is not set
@@ -171,12 +171,12 @@
 CONFIG_CAT=y
 CONFIG_DATE=y
 CONFIG_FEATURE_DATE_ISOFMT=y
-# CONFIG_FEATURE_DATE_NANO is not set
-CONFIG_FEATURE_DATE_COMPAT=y
+CONFIG_FEATURE_DATE_NANO=y
+# CONFIG_FEATURE_DATE_COMPAT is not set
 CONFIG_ID=y
 # CONFIG_GROUPS is not set
 CONFIG_TEST=y
-CONFIG_FEATURE_TEST_64=y
+# CONFIG_FEATURE_TEST_64 is not set
 CONFIG_TOUCH=y
 CONFIG_TR=y
 CONFIG_FEATURE_TR_CLASSES=y
@@ -201,7 +201,7 @@
 # CONFIG_FEATURE_DD_THIRD_STATUS_LINE is not set
 CONFIG_FEATURE_DD_IBS_OBS=y
 CONFIG_DF=y
-# CONFIG_FEATURE_DF_FANCY is not set
+CONFIG_FEATURE_DF_FANCY=y
 CONFIG_DIRNAME=y
 CONFIG_DOS2UNIX=y
 CONFIG_UNIX2DOS=y
@@ -217,7 +217,7 @@
 CONFIG_EXPR_MATH_SUPPORT_64=y
 CONFIG_FALSE=y
 CONFIG_FOLD=y
-# CONFIG_FSYNC is not set
+CONFIG_FSYNC=y
 CONFIG_HEAD=y
 CONFIG_FEATURE_FANCY_HEAD=y
 CONFIG_HOSTID=y
@@ -236,11 +236,11 @@
 CONFIG_FEATURE_LS_COLOR_IS_DEFAULT=y
 CONFIG_MD5SUM=y
 CONFIG_MKDIR=y
-CONFIG_FEATURE_MKDIR_LONG_OPTIONS=y
+# CONFIG_FEATURE_MKDIR_LONG_OPTIONS is not set
 CONFIG_MKFIFO=y
 CONFIG_MKNOD=y
 CONFIG_MV=y
-CONFIG_FEATURE_MV_LONG_OPTIONS=y
+# CONFIG_FEATURE_MV_LONG_OPTIONS is not set
 CONFIG_NICE=y
 CONFIG_NOHUP=y
 CONFIG_OD=y
@@ -264,7 +264,7 @@
 CONFIG_FEATURE_SORT_BIG=y
 # CONFIG_SPLIT is not set
 # CONFIG_FEATURE_SPLIT_FANCY is not set
-# CONFIG_STAT is not set
+CONFIG_STAT=y
 # CONFIG_FEATURE_STAT_FORMAT is not set
 CONFIG_STTY=y
 # CONFIG_SUM is not set
@@ -317,7 +317,7 @@
 CONFIG_DEALLOCVT=y
 CONFIG_DUMPKMAP=y
 # CONFIG_KBD_MODE is not set
-CONFIG_LOADFONT=y
+# CONFIG_LOADFONT is not set
 CONFIG_LOADKMAP=y
 CONFIG_OPENVT=y
 CONFIG_RESET=y
@@ -328,15 +328,11 @@
 # CONFIG_SETFONT is not set
 # CONFIG_FEATURE_SETFONT_TEXTUAL_MAP is not set
 CONFIG_DEFAULT_SETFONT_DIR=""
-CONFIG_SETKEYCODES=y
+# CONFIG_SETKEYCODES is not set
 CONFIG_SETLOGCONS=y
 # CONFIG_SHOWKEY is not set
-
-#
-# Common options for loadfont and setfont
-#
-CONFIG_FEATURE_LOADFONT_PSF2=y
-CONFIG_FEATURE_LOADFONT_RAW=y
+# CONFIG_FEATURE_LOADFONT_PSF2 is not set
+# CONFIG_FEATURE_LOADFONT_RAW is not set
 
 #
 # Debian Utilities
@@ -361,7 +357,7 @@
 CONFIG_FEATURE_VI_COLON=y
 CONFIG_FEATURE_VI_YANKMARK=y
 CONFIG_FEATURE_VI_SEARCH=y
-# CONFIG_FEATURE_VI_REGEX_SEARCH is not set
+CONFIG_FEATURE_VI_REGEX_SEARCH=y
 CONFIG_FEATURE_VI_USE_SIGNALS=y
 CONFIG_FEATURE_VI_DOT_CMD=y
 CONFIG_FEATURE_VI_READONLY=y
@@ -490,11 +486,11 @@
 # CONFIG_MODPROBE_SMALL is not set
 # CONFIG_FEATURE_MODPROBE_SMALL_OPTIONS_ON_CMDLINE is not set
 # CONFIG_FEATURE_MODPROBE_SMALL_CHECK_ALREADY_LOADED is not set
-CONFIG_INSMOD=y
-CONFIG_RMMOD=y
-CONFIG_LSMOD=y
-CONFIG_FEATURE_LSMOD_PRETTY_2_6_OUTPUT=y
-CONFIG_MODPROBE=y
+# CONFIG_INSMOD is not set
+# CONFIG_RMMOD is not set
+# CONFIG_LSMOD is not set
+# CONFIG_FEATURE_LSMOD_PRETTY_2_6_OUTPUT is not set
+# CONFIG_MODPROBE is not set
 # CONFIG_FEATURE_MODPROBE_BLACKLIST is not set
 # CONFIG_DEPMOD is not set
 
@@ -508,11 +504,11 @@
 # CONFIG_FEATURE_INSMOD_LOADINKMEM is not set
 # CONFIG_FEATURE_INSMOD_LOAD_MAP is not set
 # CONFIG_FEATURE_INSMOD_LOAD_MAP_FULL is not set
-CONFIG_FEATURE_CHECK_TAINTED_MODULE=y
-CONFIG_FEATURE_MODUTILS_ALIAS=y
-CONFIG_FEATURE_MODUTILS_SYMBOLS=y
-CONFIG_DEFAULT_MODULES_DIR="/lib/modules"
-CONFIG_DEFAULT_DEPMOD_FILE="modules.dep"
+# CONFIG_FEATURE_CHECK_TAINTED_MODULE is not set
+# CONFIG_FEATURE_MODUTILS_ALIAS is not set
+# CONFIG_FEATURE_MODUTILS_SYMBOLS is not set
+CONFIG_DEFAULT_MODULES_DIR=""
+CONFIG_DEFAULT_DEPMOD_FILE=""
 
 #
 # Linux System Utilities
@@ -529,7 +525,7 @@
 # CONFIG_FEATURE_FBSET_FANCY is not set
 # CONFIG_FEATURE_FBSET_READMODE is not set
 CONFIG_FDFLUSH=y
-CONFIG_FDFORMAT=y
+# CONFIG_FDFORMAT is not set
 # CONFIG_FDISK is not set
 # CONFIG_FDISK_SUPPORT_LARGE_DISKS is not set
 # CONFIG_FEATURE_FDISK_WRITABLE is not set
@@ -548,24 +544,24 @@
 # CONFIG_FEATURE_MINIX2 is not set
 # CONFIG_MKFS_REISER is not set
 # CONFIG_MKFS_VFAT is not set
-CONFIG_GETOPT=y
-CONFIG_FEATURE_GETOPT_LONG=y
+# CONFIG_GETOPT is not set
+# CONFIG_FEATURE_GETOPT_LONG is not set
 CONFIG_HEXDUMP=y
 # CONFIG_FEATURE_HEXDUMP_REVERSE is not set
 # CONFIG_HD is not set
-CONFIG_HWCLOCK=y
-CONFIG_FEATURE_HWCLOCK_LONG_OPTIONS=y
-CONFIG_FEATURE_HWCLOCK_ADJTIME_FHS=y
+# CONFIG_HWCLOCK is not set
+# CONFIG_FEATURE_HWCLOCK_LONG_OPTIONS is not set
+# CONFIG_FEATURE_HWCLOCK_ADJTIME_FHS is not set
 CONFIG_IPCRM=y
 CONFIG_IPCS=y
 CONFIG_LOSETUP=y
 CONFIG_LSPCI=y
 CONFIG_LSUSB=y
-CONFIG_MDEV=y
-CONFIG_FEATURE_MDEV_CONF=y
+# CONFIG_MDEV is not set
+# CONFIG_FEATURE_MDEV_CONF is not set
 # CONFIG_FEATURE_MDEV_RENAME is not set
 # CONFIG_FEATURE_MDEV_RENAME_REGEXP is not set
-CONFIG_FEATURE_MDEV_EXEC=y
+# CONFIG_FEATURE_MDEV_EXEC is not set
 # CONFIG_FEATURE_MDEV_LOAD_FIRMWARE is not set
 CONFIG_MKSWAP=y
 # CONFIG_FEATURE_MKSWAP_UUID is not set
@@ -575,7 +571,7 @@
 # CONFIG_FEATURE_MOUNT_VERBOSE is not set
 # CONFIG_FEATURE_MOUNT_HELPERS is not set
 # CONFIG_FEATURE_MOUNT_LABEL is not set
-# CONFIG_FEATURE_MOUNT_NFS is not set
+CONFIG_FEATURE_MOUNT_NFS=y
 CONFIG_FEATURE_MOUNT_CIFS=y
 CONFIG_FEATURE_MOUNT_FLAGS=y
 CONFIG_FEATURE_MOUNT_FSTAB=y
@@ -586,7 +582,7 @@
 # CONFIG_RTCWAKE is not set
 # CONFIG_SCRIPT is not set
 # CONFIG_SCRIPTREPLAY is not set
-CONFIG_SETARCH=y
+# CONFIG_SETARCH is not set
 CONFIG_SWAPONOFF=y
 # CONFIG_FEATURE_SWAPON_PRI is not set
 CONFIG_SWITCH_ROOT=y
@@ -632,8 +628,8 @@
 CONFIG_FEATURE_LESS_FLAGS=y
 # CONFIG_FEATURE_LESS_MARKS is not set
 CONFIG_FEATURE_LESS_REGEXP=y
-# CONFIG_FEATURE_LESS_WINCH is not set
-# CONFIG_FEATURE_LESS_ASK_TERMINAL is not set
+CONFIG_FEATURE_LESS_WINCH=y
+CONFIG_FEATURE_LESS_ASK_TERMINAL=y
 # CONFIG_FEATURE_LESS_DASHCMD is not set
 # CONFIG_FEATURE_LESS_LINENUMS is not set
 # CONFIG_NANDWRITE is not set
@@ -721,19 +717,19 @@
 CONFIG_NAMEIF=y
 # CONFIG_FEATURE_NAMEIF_EXTENDED is not set
 # CONFIG_NBDCLIENT is not set
-# CONFIG_NC is not set
-# CONFIG_NC_SERVER is not set
-# CONFIG_NC_EXTRA is not set
-# CONFIG_NC_110_COMPAT is not set
+CONFIG_NC=y
+CONFIG_NC_SERVER=y
+CONFIG_NC_EXTRA=y
+CONFIG_NC_110_COMPAT=y
 CONFIG_PING=y
 # CONFIG_PING6 is not set
 CONFIG_FEATURE_FANCY_PING=y
 # CONFIG_WHOIS is not set
-# CONFIG_FEATURE_IPV6 is not set
+CONFIG_FEATURE_IPV6=y
 # CONFIG_FEATURE_UNIX_LOCAL is not set
 # CONFIG_FEATURE_PREFER_IPV4_ADDRESS is not set
 # CONFIG_VERBOSE_RESOLUTION_ERRORS is not set
-# CONFIG_ARP is not set
+CONFIG_ARP=y
 CONFIG_ARPING=y
 # CONFIG_BRCTL is not set
 # CONFIG_FEATURE_BRCTL_FANCY is not set
@@ -768,22 +764,22 @@
 CONFIG_FEATURE_IFCONFIG_HW=y
 # CONFIG_FEATURE_IFCONFIG_BROADCAST_PLUS is not set
 # CONFIG_IFENSLAVE is not set
-# CONFIG_IFPLUGD is not set
-CONFIG_IFUPDOWN=y
-CONFIG_IFUPDOWN_IFSTATE_PATH="/var/run/ifstate"
-CONFIG_FEATURE_IFUPDOWN_IP=y
+CONFIG_IFPLUGD=y
+# CONFIG_IFUPDOWN is not set
+CONFIG_IFUPDOWN_IFSTATE_PATH=""
+# CONFIG_FEATURE_IFUPDOWN_IP is not set
 # CONFIG_FEATURE_IFUPDOWN_IP_BUILTIN is not set
 # CONFIG_FEATURE_IFUPDOWN_IFCONFIG_BUILTIN is not set
-CONFIG_FEATURE_IFUPDOWN_IPV4=y
+# CONFIG_FEATURE_IFUPDOWN_IPV4 is not set
 # CONFIG_FEATURE_IFUPDOWN_IPV6 is not set
-CONFIG_FEATURE_IFUPDOWN_MAPPING=y
+# CONFIG_FEATURE_IFUPDOWN_MAPPING is not set
 # CONFIG_FEATURE_IFUPDOWN_EXTERNAL_DHCP is not set
 CONFIG_INETD=y
-CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_ECHO=y
-CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_DISCARD=y
-CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_TIME=y
-CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_DAYTIME=y
-CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_CHARGEN=y
+# CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_ECHO is not set
+# CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_DISCARD is not set
+# CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_TIME is not set
+# CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_DAYTIME is not set
+# CONFIG_FEATURE_INETD_SUPPORT_BUILTIN_CHARGEN is not set
 # CONFIG_FEATURE_INETD_RPC is not set
 CONFIG_IP=y
 CONFIG_FEATURE_IP_ADDRESS=y
@@ -818,7 +814,7 @@
 # CONFIG_FEATURE_TELNETD_STANDALONE is not set
 # CONFIG_FEATURE_TELNETD_INETD_WAIT is not set
 CONFIG_TFTP=y
-# CONFIG_TFTPD is not set
+CONFIG_TFTPD=y
 
 #
 # Common options for tftp/tftpd
@@ -833,27 +829,27 @@
 # CONFIG_FEATURE_TRACEROUTE_VERBOSE is not set
 # CONFIG_FEATURE_TRACEROUTE_SOURCE_ROUTE is not set
 # CONFIG_FEATURE_TRACEROUTE_USE_ICMP is not set
-# CONFIG_TUNCTL is not set
-# CONFIG_FEATURE_TUNCTL_UG is not set
-# CONFIG_UDHCPD is not set
-# CONFIG_DHCPRELAY is not set
-# CONFIG_DUMPLEASES is not set
-# CONFIG_FEATURE_UDHCPD_WRITE_LEASES_EARLY is not set
-# CONFIG_FEATURE_UDHCPD_BASE_IP_ON_MAC is not set
-CONFIG_DHCPD_LEASES_FILE=""
+CONFIG_TUNCTL=y
+CONFIG_FEATURE_TUNCTL_UG=y
+CONFIG_UDHCPD=y
+CONFIG_DHCPRELAY=y
+CONFIG_DUMPLEASES=y
+CONFIG_FEATURE_UDHCPD_WRITE_LEASES_EARLY=y
+CONFIG_FEATURE_UDHCPD_BASE_IP_ON_MAC=y
+CONFIG_DHCPD_LEASES_FILE="/var/lib/misc/udhcpd.leases"
 CONFIG_UDHCPC=y
 CONFIG_FEATURE_UDHCPC_ARPING=y
 # CONFIG_FEATURE_UDHCP_PORT is not set
-CONFIG_UDHCP_DEBUG=0
-# CONFIG_FEATURE_UDHCP_RFC3397 is not set
+CONFIG_UDHCP_DEBUG=3
+CONFIG_FEATURE_UDHCP_RFC3397=y
 CONFIG_FEATURE_UDHCP_8021Q=y
-CONFIG_UDHCPC_DEFAULT_SCRIPT="/usr/share/udhcpc/default.script"
+CONFIG_UDHCPC_DEFAULT_SCRIPT="/etc/dhcp/udhcpc.conf"
 CONFIG_UDHCPC_SLACK_FOR_BUGGY_SERVERS=80
-CONFIG_IFUPDOWN_UDHCPC_CMD_OPTIONS="-R -n"
+CONFIG_IFUPDOWN_UDHCPC_CMD_OPTIONS=""
 # CONFIG_UDPSVD is not set
-CONFIG_VCONFIG=y
+# CONFIG_VCONFIG is not set
 CONFIG_WGET=y
-CONFIG_FEATURE_WGET_STATUSBAR=y
+# CONFIG_FEATURE_WGET_STATUSBAR is not set
 CONFIG_FEATURE_WGET_AUTHENTICATION=y
 CONFIG_FEATURE_WGET_LONG_OPTIONS=y
 CONFIG_FEATURE_WGET_TIMEOUT=y
@@ -880,12 +876,12 @@
 #
 # Process Utilities
 #
-# CONFIG_IOSTAT is not set
+CONFIG_IOSTAT=y
 # CONFIG_MPSTAT is not set
 # CONFIG_NMETER is not set
-# CONFIG_PMAP is not set
+CONFIG_PMAP=y
 # CONFIG_POWERTOP is not set
-# CONFIG_PSTREE is not set
+CONFIG_PSTREE=y
 # CONFIG_PWDX is not set
 # CONFIG_SMEMCAP is not set
 CONFIG_UPTIME=y
@@ -914,7 +910,7 @@
 # CONFIG_FEATURE_TOP_DECIMALS is not set
 # CONFIG_FEATURE_TOP_SMP_PROCESS is not set
 # CONFIG_FEATURE_TOPMEM is not set
-# CONFIG_FEATURE_SHOW_THREADS is not set
+CONFIG_FEATURE_SHOW_THREADS=y
 CONFIG_WATCH=y
 
 #
@@ -1001,13 +997,13 @@
 CONFIG_SYSLOGD=y
 CONFIG_FEATURE_ROTATE_LOGFILE=y
 CONFIG_FEATURE_REMOTE_LOG=y
-# CONFIG_FEATURE_SYSLOGD_DUP is not set
+CONFIG_FEATURE_SYSLOGD_DUP=y
 # CONFIG_FEATURE_SYSLOGD_CFG is not set
 CONFIG_FEATURE_SYSLOGD_READ_BUFFER_SIZE=256
-# CONFIG_FEATURE_IPC_SYSLOG is not set
-CONFIG_FEATURE_IPC_SYSLOG_BUFFER_SIZE=0
-# CONFIG_LOGREAD is not set
-# CONFIG_FEATURE_LOGREAD_REDUCED_LOCKING is not set
+CONFIG_FEATURE_IPC_SYSLOG=y
+CONFIG_FEATURE_IPC_SYSLOG_BUFFER_SIZE=16
+CONFIG_LOGREAD=y
+CONFIG_FEATURE_LOGREAD_REDUCED_LOCKING=y
 CONFIG_KLOGD=y
 CONFIG_FEATURE_KLOGD_KLOGCTL=y
 CONFIG_LOGGER=y
diff -Naur old/package/busybox/busybox-networking_ifplugd_iface_err.patch new/package/busybox/busybox-networking_ifplugd_iface_err.patch
--- old/package/busybox/busybox-networking_ifplugd_iface_err.patch	1969-12-31 18:00:00.000000000 -0600
+++ new/package/busybox/busybox-networking_ifplugd_iface_err.patch	2013-02-09 23:50:21.460710414 -0600
@@ -0,0 +1,23 @@
+--- busybox-1.19.3/networking/ifplugd.c	2013-01-14 22:52:56.001007445 -0600
++++ busybox-1.19.3/networking/ifplugd.c	2013-01-14 22:51:11.949004402 -0600
+@@ -685,6 +685,9 @@
+ 						G.iface_exists ? "" : "dis");
+ 				if (G.iface_exists)
+ 					maybe_up_new_iface();
++				else
++					iface_status_old = IFSTATUS_ERR;
++					/* new state:  error -> down */
+ 			}
+ 		}
+ 
+@@ -693,6 +696,7 @@
+ 		if (iface_status == IFSTATUS_ERR) {
+ 			if (!(opts & FLAG_MONITOR))
+ 				goto exiting;
++
+ 			iface_status = IFSTATUS_DOWN;
+ 		}
+ 		iface_status_str = strstatus(iface_status);
+;
+; Allow script notification of a disappeared iface, so can take action.
+;
diff -Naur old/package/Config.in new/package/Config.in
--- old/package/Config.in	2011-11-30 23:08:44.000000000 +0000
+++ new/package/Config.in	2012-12-04 17:15:15.187500000 +0000
@@ -312,12 +312,12 @@
 source "package/python-serial/Config.in"
 endmenu
 
-if BROKEN
+#if BROKEN
 menu "Java"
 source "package/java/classpath/Config.in"
 source "package/java/concierge/Config.in"
 endmenu
-endif
+#endif
 
 menu "Networking"
 source "package/libcgi/Config.in"
@@ -404,6 +404,7 @@
 source "package/can-utils/Config.in"
 source "package/ctorrent/Config.in"
 source "package/cifs-utils/Config.in"
+source "package/crda/Config.in"
 source "package/cups/Config.in"
 if BR2_PACKAGE_BUSYBOX_SHOW_OTHERS
 source "package/dhcp/Config.in"
@@ -458,6 +459,7 @@
 source "package/proftpd/Config.in"
 source "package/quagga/Config.in"
 source "package/radvd/Config.in"
+source "package/regulatory/Config.in"
 source "package/rp-pppoe/Config.in"
 source "package/rsh-redone/Config.in"
 source "package/rsync/Config.in"
@@ -537,6 +539,7 @@
 endif
 source "package/util-linux/Config.in"
 source "package/dsp-tools/Config.in"
+source "package/kmod/Config.in"
 endmenu
 
 menu "Text editors and viewers"
diff -Naur old/package/crda/Config.in new/package/crda/Config.in
--- old/package/crda/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ new/package/crda/Config.in	2012-02-13 04:35:04.562500000 +0000
@@ -0,0 +1,7 @@
+config BR2_PACKAGE_CRDA
+	bool "crda"
+	help
+	  Central Regulatory Domain Agent.
+
+	  http://linuxwireless.org/en/developers/Regulatory/CRDA
+
diff -Naur old/package/crda/crda-genl3.patch new/package/crda/crda-genl3.patch
--- old/package/crda/crda-genl3.patch	1970-01-01 01:00:00.000000000 +0100
+++ new/package/crda/crda-genl3.patch	2012-08-11 18:02:08.796875000 +0100
@@ -0,0 +1,12 @@
+diff -urN crda-1.1.2/Makefile crda-1.1.2-genl3/Makefile
+--- crda-1.1.2/Makefile	2011-08-09 19:36:19.000000000 -0400
++++ crda-1.1.2-genl3/Makefile	2012-06-24 09:50:02.043217635 -0400
+@@ -51,7 +51,7 @@
+ 
+ ifeq ($(NL3FOUND),Y)
+ CFLAGS += -DCONFIG_LIBNL30
+-NLLIBS += -lnl-genl
++NLLIBS += -lnl-genl-3
+ NLLIBNAME = libnl-3.0
+ else
+ 	ifeq ($(NL2FOUND),Y)
diff -Naur old/package/crda/crda.mk new/package/crda/crda.mk
--- old/package/crda/crda.mk	1970-01-01 01:00:00.000000000 +0100
+++ new/package/crda/crda.mk	2012-11-27 13:26:43.046875000 +0000
@@ -0,0 +1,43 @@
+#############################################################
+#
+# crda
+#
+#############################################################
+
+CRDA_VERSION = 1.1.2
+CRDA_SOURCE = crda-$(CRDA_VERSION).tar.bz2
+CRDA_SITE = http://linuxwireless.org/download/crda
+CRDA_DEPENDENCIES = host-pkg-config libnl openssl regulatory
+CRDA_CONFIG = $(CRDA_DIR)/.config
+CRDA_MAKE_ENV = PKG_CONFIG_PATH="$(STAGING_DIR)/usr/lib/pkgconfig" \
+	PKG_CONFIG="$(HOST_DIR)/usr/bin/pkg-config" \
+	GIT_DIR=$(CRDA_DIR) \
+	CFLAGS="$(TARGET_CFLAGS) -I../libnl-3.2.16/include -I../openssl-1.0.1c/include" \
+	LDFLAGS="$(TARGET_LDFLAGS) -L../../host/usr/arm-unknown-linux-gnueabi/sysroot/usr/lib -L../libnl-3.2.16/lib/.libs -L../openssl-1.0.1c" \
+	RUNTIME_PUBKEY_DIR=/usr/lib/crda/pubkeys \
+	USE_OPENSSL=1 \
+	REG_BIN=$(TARGET_DIR)/$(REGULATORY_TARGET_BINARY) \
+	NL3FOUND=Y \
+	CC="$(TARGET_CC)"
+	
+define CRDA_CONFIGURE_CMDS
+	echo "CC = $(TARGET_CC)" >$(CRDA_CONFIG)
+	echo "CFLAGS = $(TARGET_CFLAGS)" >>$(CRDA_CONFIG)
+	echo "LDFLAGS = $(TARGET_LDFLAGS)" >>$(CRDA_CONFIG)
+	cp package/crda/keys-gcrypt.c $(BUILD_DIR)/crda-$(CRDA_VERSION)
+endef
+
+define CRDA_BUILD_CMDS
+	$(CRDA_MAKE_ENV) $(MAKE) all_noverify -C $(@D)
+endef
+
+define CRDA_INSTALL_TARGET_CMDS
+	$(CRDA_MAKE_ENV) $(MAKE) -C $(@D) DESTDIR=$(TARGET_DIR) install NL3FOUND=Y
+endef
+
+define CRDA_UNINSTALL_TARGET_CMDS
+	rm -f $(TARGET_DIR)/sbin/crda
+	rm -f $(TARGET_DIR)/share/man/man8/crda.8*
+endef
+
+$(eval $(call GENTARGETS))
diff -Naur old/package/crda/keys-gcrypt.c new/package/crda/keys-gcrypt.c
--- old/package/crda/keys-gcrypt.c	1970-01-01 01:00:00.000000000 +0100
+++ new/package/crda/keys-gcrypt.c	2012-02-13 04:35:04.562500000 +0000
@@ -0,0 +1,53 @@
+static const __u8 e_0[3] = {
+	0x01, 0x00, 0x01, 
+};
+
+static const __u8 n_0[256] = {
+	0xd6, 0x57, 0x99, 0x71, 0x88, 0xd3, 0x5d, 0xce, 
+	0x2a, 0x2e, 0xad, 0x68, 0x37, 0x60, 0x24, 0xf2, 
+	0xf2, 0xcf, 0xf3, 0x93, 0xf9, 0x70, 0x54, 0xeb, 
+	0x0a, 0x9a, 0x9d, 0x48, 0x8e, 0x56, 0xe1, 0x53, 
+	0xb0, 0x75, 0xa2, 0x41, 0xe1, 0x3b, 0x5a, 0xc5, 
+	0x4d, 0x9f, 0x01, 0x37, 0x07, 0xfc, 0x0d, 0x84, 
+	0xb0, 0x4d, 0xe1, 0xd6, 0xb3, 0x9f, 0xcc, 0x8d, 
+	0x3e, 0x7c, 0x15, 0x7d, 0x0d, 0xc3, 0x8e, 0xab, 
+	0x22, 0x03, 0xfb, 0x37, 0x48, 0x2c, 0x13, 0x1b, 
+	0xf3, 0x12, 0xb9, 0x66, 0xff, 0xbb, 0xf0, 0xe9, 
+	0x01, 0x1a, 0x5b, 0x3f, 0x5b, 0x3b, 0xc3, 0x0d, 
+	0x58, 0x5d, 0x9a, 0x7d, 0x53, 0x44, 0x7b, 0xd1, 
+	0xbe, 0x62, 0x72, 0xed, 0xc5, 0xf5, 0x98, 0x24, 
+	0xf9, 0xe9, 0xc9, 0xc7, 0xad, 0x93, 0x3f, 0x64, 
+	0x16, 0x85, 0xc1, 0x57, 0xe2, 0x0f, 0xd4, 0xdc, 
+	0xaa, 0xa8, 0xac, 0xc8, 0x5d, 0x6c, 0xa8, 0x4e, 
+	0x7e, 0x70, 0xc1, 0xd3, 0x59, 0x16, 0xf1, 0x73, 
+	0x5f, 0xab, 0xb6, 0x88, 0xeb, 0xd1, 0xe1, 0x98, 
+	0xac, 0xd7, 0xca, 0xdb, 0x46, 0xb0, 0x94, 0x3e, 
+	0x4a, 0x56, 0x3e, 0x84, 0x8f, 0x3a, 0x2d, 0xaf, 
+	0xd1, 0x03, 0x58, 0x0b, 0x0b, 0xd5, 0xb1, 0xff, 
+	0x90, 0x67, 0xcc, 0x76, 0x7a, 0x6e, 0xb5, 0xcc, 
+	0x8c, 0x6a, 0x92, 0xd0, 0xa5, 0x15, 0x6b, 0xb8, 
+	0xef, 0xe3, 0x11, 0xd8, 0xd2, 0x9a, 0x10, 0x04, 
+	0xfe, 0xc8, 0xf5, 0xb6, 0x61, 0xdc, 0x01, 0x70, 
+	0x53, 0xea, 0x3e, 0x5a, 0x97, 0x83, 0x9a, 0x2e, 
+	0x20, 0x5a, 0x5a, 0x73, 0xfe, 0xfa, 0xbb, 0xa7, 
+	0x7b, 0xba, 0xe3, 0x8a, 0x6c, 0x1f, 0xce, 0x9d, 
+	0x39, 0xe3, 0xdf, 0xeb, 0xba, 0x94, 0x14, 0x33, 
+	0xb9, 0xc0, 0x65, 0x10, 0xcb, 0xcb, 0x35, 0xe3, 
+	0xdc, 0x9f, 0xca, 0x11, 0xc8, 0xba, 0x75, 0x7b, 
+	0x63, 0xa2, 0x70, 0x54, 0x16, 0xa0, 0xd8, 0xe1, 
+};
+
+
+struct key_params {
+	const __u8 *e, *n;
+	__u32 len_e, len_n; 
+};
+
+#define KEYS(_e, _n) {			\
+	.e = _e, .len_e = sizeof(_e),	\
+	.n = _n, .len_n = sizeof(_n),	\
+}
+
+static const struct key_params keys[] = {
+	KEYS(e_0, n_0),
+};
diff -Naur old/package/dhcp/dhclient-script new/package/dhcp/dhclient-script
--- old/package/dhcp/dhclient-script	2011-11-30 23:08:44.000000000 +0000
+++ new/package/dhcp/dhclient-script	2012-02-07 15:26:17.296875000 +0000
@@ -1,2 +1,318 @@
 #!/bin/sh
-# empty dhclient-script to keep dhclient from complaining
+# dhclient-script for Linux. Dan Halbert, March, 1997.
+# Updated for Linux 2.[12] by Brian J. Murrell, January 1999.
+# No guarantees about this. I'm a novice at the details of Linux
+# networking.
+
+# Notes:
+
+# 0. This script is based on the netbsd script supplied with dhcp-970306.
+
+# 1. ifconfig down apparently deletes all relevant routes and flushes
+# the arp cache, so this doesn't need to be done explicitly.
+
+# 2. The alias address handling here has not been tested AT ALL.
+# I'm just going by the doc of modern Linux ip aliasing, which uses
+# notations like eth0:0, eth0:1, for each alias.
+
+# 3. I have to calculate the network address, and calculate the broadcast
+# address if it is not supplied. This might be much more easily done
+# by the dhclient C code, and passed on.
+
+# 4. TIMEOUT not tested. ping has a flag I don't know, and I'm suspicious
+# of the $1 in its args.
+
+# 'ip' just looks too weird.  /sbin/ip looks less weird.
+ip=/sbin/ip
+
+make_resolv_conf() {
+  if [ x"$new_domain_name_servers" != x ]; then
+    cat /dev/null > /etc/resolv.conf.dhclient
+    chmod 644 /etc/resolv.conf.dhclient
+    if [ x"$new_domain_search" != x ]; then
+      echo search $new_domain_search >> /etc/resolv.conf.dhclient
+    elif [ x"$new_domain_name" != x ]; then
+      # Note that the DHCP 'Domain Name Option' is really just a domain
+      # name, and that this practice of using the domain name option as
+      # a search path is both nonstandard and deprecated.
+      echo search $new_domain_name >> /etc/resolv.conf.dhclient
+    fi
+    for nameserver in $new_domain_name_servers; do
+      echo nameserver $nameserver >>/etc/resolv.conf.dhclient
+    done
+
+    mv /etc/resolv.conf.dhclient /etc/resolv.conf
+  elif [ "x${new_dhcp6_name_servers}" != x ] ; then
+    cat /dev/null > /etc/resolv.conf.dhclient6
+    chmod 644 /etc/resolv.conf.dhclient6
+
+    if [ "x${new_dhcp6_domain_search}" != x ] ; then
+      echo search ${new_dhcp6_domain_search} >> /etc/resolv.conf.dhclient6
+    fi
+    shopt -s nocasematch 
+    for nameserver in ${new_dhcp6_name_servers} ; do
+      # If the nameserver has a link-local address
+      # add a <zone_id> (interface name) to it.
+      if  [[ "$nameserver" =~ ^fe80:: ]]
+      then
+	zone_id="%$interface"
+      else
+	zone_id=
+      fi
+      echo nameserver ${nameserver}$zone_id >> /etc/resolv.conf.dhclient6
+    done
+    shopt -u nocasematch 
+
+    mv /etc/resolv.conf.dhclient6 /etc/resolv.conf
+  fi
+}
+
+# Must be used on exit.   Invokes the local dhcp client exit hooks, if any.
+exit_with_hooks() {
+  exit_status=$1
+  if [ -f /etc/dhclient-exit-hooks ]; then
+    . /etc/dhclient-exit-hooks
+  fi
+# probably should do something with exit status of the local script
+  exit $exit_status
+}
+
+# Invoke the local dhcp client enter hooks, if they exist.
+if [ -f /etc/dhclient-enter-hooks ]; then
+  exit_status=0
+  . /etc/dhclient-enter-hooks
+  # allow the local script to abort processing of this state
+  # local script must set exit_status variable to nonzero.
+  if [ $exit_status -ne 0 ]; then
+    exit $exit_status
+  fi
+fi
+
+###
+### DHCPv4 Handlers
+###
+
+if [ x$new_broadcast_address != x ]; then
+  new_broadcast_arg="broadcast $new_broadcast_address"
+fi
+if [ x$old_broadcast_address != x ]; then
+  old_broadcast_arg="broadcast $old_broadcast_address"
+fi
+if [ x$new_subnet_mask != x ]; then
+  new_subnet_arg="netmask $new_subnet_mask"
+fi
+if [ x$old_subnet_mask != x ]; then
+  old_subnet_arg="netmask $old_subnet_mask"
+fi
+if [ x$alias_subnet_mask != x ]; then
+  alias_subnet_arg="netmask $alias_subnet_mask"
+fi
+if [ x$new_interface_mtu != x ]; then
+  mtu_arg="mtu $new_interface_mtu"
+fi
+if [ x$IF_METRIC != x ]; then
+  metric_arg="metric $IF_METRIC"
+fi
+
+if [ x$reason = xMEDIUM ]; then
+  # Linux doesn't do mediums (ok, ok, media).
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xPREINIT ]; then
+  if [ x$alias_ip_address != x ]; then
+    # Bring down alias interface. Its routes will disappear too.
+    ifconfig $interface:0- inet 0
+  fi
+  ifconfig $interface 0 up
+
+  # We need to give the kernel some time to get the interface up.
+  sleep 1
+
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xARPCHECK ] || [ x$reason = xARPSEND ]; then
+  exit_with_hooks 0
+fi
+  
+if [ x$reason = xBOUND ] || [ x$reason = xRENEW ] || \
+   [ x$reason = xREBIND ] || [ x$reason = xREBOOT ]; then
+  current_hostname=`hostname`
+  if [ x$current_hostname = x ] || \
+     [ x$current_hostname = "x(none)" ] || \
+     [ x$current_hostname = xlocalhost ] || \
+     [ x$current_hostname = x$old_host_name ]; then
+    if [ x$new_host_name != x$old_host_name ]; then
+      hostname "$new_host_name"
+    fi
+  fi
+    
+  if [ x$old_ip_address != x ] && [ x$alias_ip_address != x ] && \
+		[ x$alias_ip_address != x$old_ip_address ]; then
+    # Possible new alias. Remove old alias.
+    ifconfig $interface:0- inet 0
+  fi
+  if [ x$old_ip_address != x ] && [ x$old_ip_address != x$new_ip_address ]; then
+    # IP address changed. Bringing down the interface will delete all routes,
+    # and clear the ARP cache.
+    ifconfig $interface inet 0 down
+
+  fi
+  if [ x$old_ip_address = x ] || [ x$old_ip_address != x$new_ip_address ] || \
+     [ x$reason = xBOUND ] || [ x$reason = xREBOOT ]; then
+
+    ifconfig $interface inet $new_ip_address $new_subnet_arg \
+					$new_broadcast_arg $mtu_arg
+    # Add a network route to the computed network address.
+    for router in $new_routers; do
+      if [ "x$new_subnet_mask" = "x255.255.255.255" ] ; then
+	route add -host $router dev $interface
+      fi
+      route add default gw $router $metric_arg dev $interface
+    done
+  else
+    # we haven't changed the address, have we changed other options           
+    # that we wish to update?
+    if [ x$new_routers != x ] && [ x$new_routers != x$old_routers ] ; then
+      # if we've changed routers delete the old and add the new.
+      for router in $old_routers; do
+        route del default gw $router
+      done
+      for router in $new_routers; do
+        if [ "x$new_subnet_mask" = "x255.255.255.255" ] ; then
+	  route add -host $router dev $interface
+	fi
+	route add default gw $router $metric_arg dev $interface
+      done
+    fi
+  fi
+  if [ x$new_ip_address != x$alias_ip_address ] && [ x$alias_ip_address != x ];
+   then
+    ifconfig $interface:0- inet 0
+    ifconfig $interface:0 inet $alias_ip_address $alias_subnet_arg
+    route add -host $alias_ip_address $interface:0
+  fi
+  make_resolv_conf
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xEXPIRE ] || [ x$reason = xFAIL ] || [ x$reason = xRELEASE ] \
+   || [ x$reason = xSTOP ]; then
+  if [ x$alias_ip_address != x ]; then
+    # Turn off alias interface.
+    ifconfig $interface:0- inet 0
+  fi
+  if [ x$old_ip_address != x ]; then
+    # Shut down interface, which will delete routes and clear arp cache.
+    ifconfig $interface inet 0 down
+  fi
+  if [ x$alias_ip_address != x ]; then
+    ifconfig $interface:0 inet $alias_ip_address $alias_subnet_arg
+    route add -host $alias_ip_address $interface:0
+  fi
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xTIMEOUT ]; then
+  if [ x$alias_ip_address != x ]; then
+    ifconfig $interface:0- inet 0
+  fi
+  ifconfig $interface inet $new_ip_address $new_subnet_arg \
+					$new_broadcast_arg $mtu_arg
+  set $new_routers
+  if ping -q -c 1 $1; then
+    if [ x$new_ip_address != x$alias_ip_address ] && \
+			[ x$alias_ip_address != x ]; then
+      ifconfig $interface:0 inet $alias_ip_address $alias_subnet_arg
+      route add -host $alias_ip_address dev $interface:0
+    fi
+    for router in $new_routers; do
+      if [ "x$new_subnet_mask" = "x255.255.255.255" ] ; then
+	route add -host $router dev $interface
+      fi
+      route add default gw $router $metric_arg dev $interface
+    done
+    make_resolv_conf
+    exit_with_hooks 0
+  fi
+  ifconfig $interface inet 0 down
+  exit_with_hooks 1
+fi
+
+###
+### DHCPv6 Handlers
+###
+
+if [ x$reason = xPREINIT6 ] ; then
+  # Ensure interface is up.
+  ${ip} link set ${interface} up
+
+  # Remove any stale addresses from aborted clients.
+  ${ip} -f inet6 addr flush dev ${interface} scope global permanent
+
+  exit_with_hooks 0
+fi
+
+if [ x${old_ip6_prefix} != x ] || [ x${new_ip6_prefix} != x ] ; then
+    echo Prefix ${reason} old=${old_ip6_prefix} new=${new_ip6_prefix}
+
+    exit_with_hooks 0
+fi
+
+if [ x$reason = xBOUND6 ] ; then
+  if [ x${new_ip6_address} = x ] || [ x${new_ip6_prefixlen} = x ] ; then
+    exit_with_hooks 2;
+  fi
+
+  ${ip} -f inet6 addr add ${new_ip6_address}/${new_ip6_prefixlen} \
+	dev ${interface} scope global
+
+  # Check for nameserver options.
+  make_resolv_conf
+
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xRENEW6 ] || [ x$reason = xREBIND6 ] ; then
+  if [ x${new_ip6_address} = x ] || [ x${new_ip6_prefixlen} = x ] ; then
+    exit_with_hooks 2;
+  fi
+
+  ${ip} -f inet6 addr add ${new_ip6_address}/${new_ip6_prefixlen} \
+	dev ${interface} scope global
+
+  # Make sure nothing has moved around on us.
+
+  # Nameservers/domains/etc.
+  if [ "x${new_dhcp6_name_servers}" != "x${old_dhcp6_name_servers}" ] ||
+     [ "x${new_dhcp6_domain_search}" != "x${old_dhcp6_domain_search}" ] ; then
+    make_resolv_conf
+  fi
+
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xDEPREF6 ] ; then
+  if [ x${new_ip6_prefixlen} = x ] ; then
+    exit_with_hooks 2;
+  fi
+
+  ${ip} -f inet6 addr change ${new_ip6_address}/${new_ip6_prefixlen} \
+       dev ${interface} scope global preferred_lft 0
+
+  exit_with_hooks 0
+fi
+
+if [ x$reason = xEXPIRE6 -o x$reason = xRELEASE6 -o x$reason = xSTOP6 ] ; then
+  if [ x${old_ip6_address} = x ] || [ x${old_ip6_prefixlen} = x ] ; then
+    exit_with_hooks 2;
+  fi
+
+  ${ip} -f inet6 addr del ${old_ip6_address}/${old_ip6_prefixlen} \
+	dev ${interface}
+
+  exit_with_hooks 0
+fi
+
+exit_with_hooks 0
diff -Naur old/package/dhcp/dhcp.mk new/package/dhcp/dhcp.mk
--- old/package/dhcp/dhcp.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/dhcp/dhcp.mk	2011-12-16 02:51:39.281250000 +0000
@@ -4,7 +4,7 @@
 #
 #############################################################
 
-DHCP_VERSION  = 4.1-ESV-R3
+DHCP_VERSION  = 4.1-ESV-R4
 DHCP_SITE     = http://ftp.isc.org/isc/dhcp
 DHCP_CONF_ENV = ac_cv_file__dev_random=yes
 DHCP_CONF_OPT = \
@@ -25,7 +25,7 @@
 	(cd $(TARGET_DIR)/var/lib; ln -snf /tmp dhcp)
 	$(INSTALL) -m 0755 -D $(@D)/server/dhcpd $(TARGET_DIR)/usr/sbin/dhcpd
 	$(INSTALL) -m 0755 -D package/dhcp/S80dhcp-server \
-		$(TARGET_DIR)/etc/init.d/S80dhcp-server
+		$(TARGET_DIR)/etc/init.d/opt/S80dhcp-server
 	$(INSTALL) -m 0644 -D package/dhcp/dhcpd.conf \
 		$(TARGET_DIR)/etc/dhcp/dhcpd.conf
 endef
@@ -38,7 +38,7 @@
 	$(INSTALL) -m 0755 -D $(DHCP_DIR)/relay/dhcrelay \
 		$(TARGET_DIR)/usr/sbin/dhcrelay
 	$(INSTALL) -m 0755 -D package/dhcp/S80dhcp-relay \
-		$(TARGET_DIR)/etc/init.d/S80dhcp-relay
+		$(TARGET_DIR)/etc/init.d/opt/S80dhcp-relay
 endef
 endif
 
diff -Naur old/package/ebtables/Config.in new/package/ebtables/Config.in
--- old/package/ebtables/Config.in	2011-11-30 23:08:44.000000000 +0000
+++ new/package/ebtables/Config.in	2011-09-20 03:43:28.406250000 +0100
@@ -1,10 +1,6 @@
 config BR2_PACKAGE_EBTABLES
 	bool "ebtables"
-	depends on BR2_INET_IPV6
 	help
-	  Ethernet bridge frame table administration
+	  A filtering tool for a Linux-based bridging firewall.
 
-	  http://ebtables.sourceforge.net
-
-comment "ebtables requires a toolchain with IPv6 support"
-	depends on !BR2_INET_IPV6
+	  http://ebtables.sourceforge.net/
diff -Naur old/package/ebtables/ebtables-add-arpnat-target.patch new/package/ebtables/ebtables-add-arpnat-target.patch
--- old/package/ebtables/ebtables-add-arpnat-target.patch	1970-01-01 01:00:00.000000000 +0100
+++ new/package/ebtables/ebtables-add-arpnat-target.patch	2011-12-15 20:44:48.468750000 +0000
@@ -0,0 +1,132 @@
+diff -urN ebtables-v2.0.10-2/Makefile ebtables-v2.0.10-2-arpnat/Makefile
+--- ebtables-v2.0.10-2/Makefile	2011-08-11 14:56:16.000000000 -0400
++++ ebtables-v2.0.10-2-arpnat/Makefile	2011-09-18 18:49:20.071816739 -0400
+@@ -17,9 +17,9 @@
+ SYSCONFIGDIR:=/etc/sysconfig
+ DESTDIR:=
+ 
+-CFLAGS:=-Wall -Wunused -Werror
++#CFLAGS:=-Wall -Wunused -Werror
+ CFLAGS_SH_LIB:=-fPIC -O3
+-CC:=gcc
++#CC:=gcc
+ 
+ ifeq ($(shell uname -m),sparc64)
+ CFLAGS+=-DEBT_MIN_ALIGN=8 -DKERNEL_64_USERSPACE_32
+diff -urN ebtables-v2.0.10-2/extensions/ebt_nat.c ebtables-v2.0.10-2-arpnat/extensions/ebt_nat.c
+--- ebtables-v2.0.10-2/extensions/ebt_nat.c	2011-08-11 14:56:16.000000000 -0400
++++ ebtables-v2.0.10-2-arpnat/extensions/ebt_nat.c	2011-09-17 12:59:35.497511502 -0400
+@@ -21,6 +21,7 @@
+ #define NAT_S_TARGET '2'
+ #define NAT_D_TARGET '2'
+ #define NAT_S_ARP '3'
++#define NAT_ARP_TARGET '2'
+ static struct option opts_s[] =
+ {
+ 	{ "to-source"     , required_argument, 0, NAT_S },
+@@ -30,6 +31,12 @@
+ 	{ 0 }
+ };
+ 
++static struct option opts_arpnat[] =
++{
++	{ "arpnat-target" , required_argument, 0, NAT_ARP_TARGET },
++	{ 0 }
++};
++
+ static struct option opts_d[] =
+ {
+ 	{ "to-destination", required_argument, 0, NAT_D },
+@@ -47,6 +54,13 @@
+ 	" --snat-arp             : also change src address in arp msg\n");
+ }
+ 
++static void print_help_arpnat()
++{
++	printf(
++	"arpnat options:\n"
++	" --arpnat-target target   : ACCEPT, DROP, RETURN or CONTINUE\n");
++}
++
+ static void print_help_d()
+ {
+ 	printf(
+@@ -55,6 +69,54 @@
+ 	" --dnat-target target   : ACCEPT, DROP, RETURN or CONTINUE\n");
+ }
+ 
++static void init_arpnat(struct ebt_entry_target *target)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
++
++   natinfo->target = EBT_ACCEPT;
++   return;
++}
++#define OPT_ARPNAT_TARGET 0x2
++static int parse_arpnat(int c, char **argv, int argc,
++   const struct ebt_u_entry *entry, unsigned int *flags,
++   struct ebt_entry_target **target)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)(*target)->data;
++
++   switch (c) {
++   case NAT_ARP_TARGET:
++       ebt_check_option2(flags, OPT_ARPNAT_TARGET);
++       if (FILL_TARGET(optarg, natinfo->target))
++           ebt_print_error2("Illegal --arpnat-target target");
++       break;
++   default:
++       return 0;
++   }
++   return 1;
++}
++
++static void final_check_arpnat(const struct ebt_u_entry *entry,
++   const struct ebt_entry_target *target, const char *name,
++   unsigned int hookmask, unsigned int time)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
++
++   if (BASE_CHAIN && natinfo->target == EBT_RETURN)
++       ebt_print_error("--arpnat-target RETURN not allowed on base chain");
++   CLEAR_BASE_CHAIN_BIT;
++   if (((hookmask & ~((1 << NF_BR_PRE_ROUTING) | (1 << NF_BR_POST_ROUTING)))
++      || strcmp(name, "nat")))
++       ebt_print_error("Wrong chain for arpnat");
++}
++
++static void print_arpnat(const struct ebt_u_entry *entry,
++   const struct ebt_entry_target *target)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
++
++   printf(" --arpnat-target %s", TARGET_NAME(natinfo->target));
++}
++
+ static void init_s(struct ebt_entry_target *target)
+ {
+ 	struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
+@@ -231,8 +293,23 @@
+ 	.extra_ops	= opts_d,
+ };
+ 
++#define EBT_ARPNAT_TARGET "arpnat"
++static struct ebt_u_target arpnat_target =
++{
++   .name       = EBT_ARPNAT_TARGET,
++   .size       = sizeof(struct ebt_nat_info),
++   .help       = print_help_arpnat,
++   .init       = init_arpnat,
++   .parse      = parse_arpnat,
++   .final_check    = final_check_arpnat,
++   .print      = print_arpnat,
++   .compare    = compare,
++   .extra_ops  = opts_arpnat,
++};
++
+ void _init(void)
+ {
+ 	ebt_register_target(&snat_target);
+ 	ebt_register_target(&dnat_target);
++	ebt_register_target(&arpnat_target);
+ }
diff -Naur old/package/ebtables/ebtables.mk new/package/ebtables/ebtables.mk
--- old/package/ebtables/ebtables.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/ebtables/ebtables.mk	2012-01-25 17:52:11.281250000 +0000
@@ -1,30 +1,24 @@
 #############################################################
 #
-# ebtables
+# ebtables - A filtering tool for a Linux-based bridging firewall.
 #
 #############################################################
-EBTABLES_VERSION = 2.0.10-1
+EBTABLES_VERSION = 2.0.10-4
 EBTABLES_SOURCE = ebtables-v$(EBTABLES_VERSION).tar.gz
 EBTABLES_SITE = http://$(BR2_SOURCEFORGE_MIRROR).dl.sourceforge.net/sourceforge/ebtables
+EBTABLES_AUTORECONF = NO
+EBTABLES_INSTALL_STAGING = NO
+EBTABLES_INSTALL_TARGET = YES
 
 define EBTABLES_BUILD_CMDS
-	$(MAKE) $(TARGET_CONFIGURE_OPTS) LIBDIR=/lib/ebtables -C $(@D)
+	$(MAKE) CC="$(TARGET_CC)" LD="$(TARGET_LD)" -C $(@D) all
 endef
 
 define EBTABLES_INSTALL_TARGET_CMDS
-	for so in $(@D)/$(EBTABLES_SUBDIR)/*.so \
-		$(@D)/$(EBTABLES_SUBDIR)/extensions/*.so; \
-		do \
-		$(INSTALL) -m 0755 -D $${so} \
-			$(TARGET_DIR)/lib/ebtables/`basename $${so}`; \
-	done
-	$(INSTALL) -m 0755 -D $(@D)/$(EBTABLES_SUBDIR)/ebtables \
-		$(TARGET_DIR)/sbin/ebtables
-endef
-
-define EBTABLES_UNINSTALL_TARGET_CMDS
-	rm -rf $(TARGET_DIR)/lib/ebtables
-	rm -f $(TARGET_DIR)/sbin/ebtables
+	$(INSTALL) -D -m 0755 $(@D)/ebtables $(TARGET_DIR)/sbin
+	$(INSTALL) -D -m 0644 $(@D)/ethertypes $(TARGET_DIR)/etc
+	$(INSTALL) -D -m 0755 $(@D)/libebtc.so $(TARGET_DIR)/usr/lib
+	$(INSTALL) -D -m 0755 $(@D)/extensions/libebt*.so $(TARGET_DIR)/usr/lib
 endef
 
 $(eval $(call GENTARGETS))
diff -Naur old/package/input-event-daemon/input-event-daemon.mk new/package/input-event-daemon/input-event-daemon.mk
--- old/package/input-event-daemon/input-event-daemon.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/input-event-daemon/input-event-daemon.mk	2012-02-01 22:08:50.093750000 +0000
@@ -19,9 +19,9 @@
 	[ -f $(TARGET_DIR)/etc/input-event-daemon.conf ] || \
 		$(INSTALL) -m 644 -D $(@D)/docs/sample.conf \
 			$(TARGET_DIR)/etc/input-event-daemon.conf
-	[ -f $(TARGET_DIR)/etc/init.d/S99input-event-daemon ] || \
+	[ -f $(TARGET_DIR)/etc/init.d/opt/S99input-event-daemon ] || \
 		$(INSTALL) -m 0755 -D package/input-event-daemon/S99input-event-daemon \
-			$(TARGET_DIR)/etc/init.d/S99input-event-daemon
+			$(TARGET_DIR)/etc/init.d/opt/S99input-event-daemon
 endef
 
 define INPUT_EVENT_DAEMON_CLEAN_CMDS
diff -Naur old/package/iw/iw-libnl-30-support.patch new/package/iw/iw-libnl-30-support.patch
--- old/package/iw/iw-libnl-30-support.patch	2011-11-30 23:08:44.000000000 +0000
+++ new/package/iw/iw-libnl-30-support.patch	2012-08-11 18:00:31.671875000 +0100
@@ -20,8 +20,8 @@
 
 +ifeq ($(NL3FOUND),Y)
 +CFLAGS += -DCONFIG_LIBNL20
-+LIBS += -lnl-genl
-+NLLIBNAME = libnl-3.0
++LIBS +=
++NLLIBNAME = libnl-genl-3.0
 +endif
 +
  ifeq ($(NLLIBNAME),)
diff -Naur old/package/kmod/Config.in new/package/kmod/Config.in
--- old/package/kmod/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ new/package/kmod/Config.in	2012-08-12 15:54:38.203125000 +0100
@@ -0,0 +1,14 @@
+config BR2_PACKAGE_KMOD
+	bool "kmod"
+	help
+	  handle kernel modules
+
+	  http://git.kernel.org/?p=utils/kernel/kmod/kmod.git
+
+config BR2_PACKAGE_KMOD_TOOLS
+	bool "kmod utilities"
+	depends on BR2_PACKAGE_KMOD
+	depends on BR2_PACKAGE_BUSYBOX_SHOW_OTHERS
+	help
+	  Install kmod module utilities (depmod, insmod, lsmod,
+	  modinfo, modprobe, rmmod).
diff -Naur old/package/kmod/kmod.mk new/package/kmod/kmod.mk
--- old/package/kmod/kmod.mk	1970-01-01 01:00:00.000000000 +0100
+++ new/package/kmod/kmod.mk	2012-08-12 15:54:38.203125000 +0100
@@ -0,0 +1,33 @@
+KMOD_VERSION = 9
+KMOD_SOURCE = kmod-$(KMOD_VERSION).tar.xz
+KMOD_SITE = $(BR2_KERNEL_MIRROR)/linux/utils/kernel/kmod/
+KMOD_INSTALL_STAGING = YES
+KMOD_DEPENDENCIES = host-pkg-config
+
+ifeq ($(BR2_PACKAGE_ZLIB),y)
+KMOD_DEPENDENCIES += zlib
+KMOD_CONF_OPT += --with-zlib
+endif
+
+ifeq ($(BR2_PACKAGE_XZ),y)
+KMOD_DEPENDENCIES += xz
+KMOD_CONF_OPT += --with-xz
+endif
+
+ifeq ($(BR2_PACKAGE_KMOD_TOOLS),y)
+# take precedence over busybox / module-init-tools implementations
+KMOD_DEPENDENCIES += \
+	$(if $(BR2_PACKAGE_BUSYBOX),busybox) \
+	$(if $(BR2_PACKAGE_MODULE_INIT_TOOLS),module-init-tools)
+
+define KMOD_INSTALL_TOOLS
+	for i in depmod insmod lsmod modinfo modprobe rmmod; \
+	do ln -sf ../usr/bin/kmod $(TARGET_DIR)/sbin/$$i; done
+endef
+
+KMOD_POST_INSTALL_TARGET_HOOKS += KMOD_INSTALL_TOOLS
+else
+KMOD_CONF_OPT += --disable-tools
+endif
+
+$(eval $(call AUTOTARGETS))
diff -Naur old/package/libnl/libnl.mk new/package/libnl/libnl.mk
--- old/package/libnl/libnl.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/libnl/libnl.mk	2012-11-27 13:26:43.046875000 +0000
@@ -4,7 +4,7 @@
 #
 #############################################################
 
-LIBNL_VERSION = 3.0
+LIBNL_VERSION = 3.2.16
 LIBNL_SOURCE = libnl-$(LIBNL_VERSION).tar.gz
 LIBNL_SITE = http://www.infradead.org/~tgr/libnl/files/
 LIBNL_INSTALL_STAGING = YES
diff -Naur old/package/lockfile-progs/lockfile-progs.mk new/package/lockfile-progs/lockfile-progs.mk
--- old/package/lockfile-progs/lockfile-progs.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/lockfile-progs/lockfile-progs.mk	2011-12-14 03:58:49.265625000 +0000
@@ -12,6 +12,8 @@
 	$(addprefix lockfile-,check create remove touch) \
 	$(addprefix mail-,lock touchlock unlock)
 
+LOCKFILE_PROGS_DEPENDENCIES = liblockfile
+
 define LOCKFILE_PROGS_BUILD_CMDS
 	$(MAKE) $(TARGET_CONFIGURE_OPTS) -C $(@D)
 endef
diff -Naur old/package/openssh/openssh.mk new/package/openssh/openssh.mk
--- old/package/openssh/openssh.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/openssh/openssh.mk	2012-05-25 17:28:37.062500000 +0100
@@ -8,12 +8,13 @@
 OPENSSH_SITE = http://ftp.openbsd.org/pub/OpenBSD/OpenSSH/portable
 OPENSSH_CONF_ENV = LD="$(TARGET_CC)" LDFLAGS="$(TARGET_CFLAGS)"
 OPENSSH_CONF_OPT = --libexecdir=/usr/lib --disable-lastlog --disable-utmp \
-		--disable-utmpx --disable-wtmp --disable-wtmpx --disable-strip
+		--disable-utmpx --disable-wtmp --disable-wtmpx --disable-strip \
+		--sysconfdir=/etc/ssh
 
 OPENSSH_DEPENDENCIES = zlib openssl
 
 define OPENSSH_INSTALL_INITSCRIPT
-	$(INSTALL) -D -m 755 package/openssh/S50sshd $(TARGET_DIR)/etc/init.d/S50sshd
+	$(INSTALL) -D -m 755 package/openssh/S50sshd $(TARGET_DIR)/etc/init.d/opt/S50sshd
 endef
 
 OPENSSH_POST_INSTALL_TARGET_HOOKS += OPENSSH_INSTALL_INITSCRIPT
diff -Naur old/package/openssh/S50sshd new/package/openssh/S50sshd
--- old/package/openssh/S50sshd	2011-11-30 23:08:44.000000000 +0000
+++ new/package/openssh/S50sshd	2012-05-25 17:28:37.062500000 +0100
@@ -2,70 +2,88 @@
 #
 # sshd        Starts sshd.
 #
-
-# Make sure the ssh-keygen progam exists
-[ -f /usr/bin/ssh-keygen ] || exit 0
-
-# Check for the SSH1 RSA key
-if [ ! -f /etc/ssh_host_key ] ; then
-	echo Generating RSA Key...
-	/usr/bin/ssh-keygen -t rsa1 -f /etc/ssh_host_key -C '' -N ''
-fi
-
-# Check for the SSH2 RSA key
-if [ ! -f /etc/ssh_host_rsa_key ] ; then
-	echo Generating RSA Key...
-	/usr/bin/ssh-keygen -t rsa -f /etc/ssh_host_rsa_key -C '' -N ''
-fi
-
-# Check for the SSH2 DSA key
-if [ ! -f /etc/ssh_host_dsa_key ] ; then
-	echo Generating DSA Key...
-	echo THIS CAN TAKE A MINUTE OR TWO DEPENDING ON YOUR PROCESSOR!
-	echo
-        /usr/bin/ssh-keygen -t dsa -f /etc/ssh_host_dsa_key -C '' -N ''
-fi
-
-# Check for the SSH2 ECDSA key
-if [ ! -f /etc/ssh_host_ecdsa_key ]; then
-	echo Generating ECDSA Key...
-	echo THIS CAN TAKE A MINUTE OR TWO DEPENDING ON YOUR PROCESSOR!
-	echo
-	/usr/bin/ssh-keygen -t ecdsa -f /etc/ssh_host_ecdsa_key -C '' -N ''
-fi
-                
 umask 077
 
+check() {
+  echo "Checking sshd configuration..."
+  
+  # Make sure the ssh-keygen progam exists
+  [ -f /usr/bin/ssh-keygen ] || exit 0
+
+  # Check that the sshd_config file exists
+  [ -f /etc/ssh/sshd_config ] || exit 0
+
+  # See if need to generate the SSH1 RSA key
+  if [ ! -f /etc/ssh/ssh_host_key ] \
+  && grep -q "^HostKey.*host_key" /etc/ssh/sshd_config
+  then
+    echo Generating RSA Key...
+    /usr/bin/ssh-keygen -t rsa1 -f /etc/ssh/ssh_host_key -C '' -N ''
+  fi
+
+  # See if need to generate the SSH2 RSA key
+  if [ ! -f /etc/ssh/ssh_host_rsa_key ] \
+  && grep -q "^HostKey.*host_rsa_key" /etc/ssh/sshd_config
+  then
+    echo Generating RSA Key...
+    /usr/bin/ssh-keygen -t rsa -f /etc/ssh/ssh_host_rsa_key -C '' -N ''
+  fi
+
+  # See if need to generate the SSH2 DSA key
+  if [ ! -f /etc/ssh/ssh_host_dsa_key ] \
+  && grep -q "^HostKey.*host_dsa_key" /etc/ssh/sshd_config
+  then
+    echo Generating DSA Key... \(may take awhile\)
+    /usr/bin/ssh-keygen -t dsa -f /etc/ssh/ssh_host_dsa_key -C '' -N ''
+  fi
+
+  # See if need to generate the SSH2 ECDSA key
+  if [ ! -f /etc/ssh/ssh_host_ecdsa_key ] \
+  && grep -q "^HostKey.*host_ecdsa_key" /etc/ssh/sshd_config
+  then
+    echo Generating ECDSA Key... \(may take awhile\)
+    /usr/bin/ssh-keygen -t ecdsa -f /etc/ssh/ssh_host_ecdsa_key -C '' -N ''
+  fi
+}                
+
 start() {
- 	echo -n "Starting sshd: "
-	/usr/sbin/sshd
-	touch /var/lock/sshd
-	echo "OK"
-}	
+ echo -n "Starting sshd: "
+  /usr/sbin/sshd
+  #/usr/sbin/sshd -f /etc/ssh/sshd_config
+# touch /var/lock/sshd.lock
+  echo "OK"
+}  
 stop() {
-	echo -n "Stopping sshd: "
-        killall	sshd 
-	rm -f /var/lock/sshd
-	echo "OK" 
+  echo -n "Stopping sshd: "
+  killall sshd 2>/dev/null
+# rm -f /var/lock/sshd.lock
+  echo "OK" 
 }
 restart() {
-	stop
-	start
-}	
+  stop
+  check
+  start
+}  
 
 case "$1" in
+  check)
+    # support inetd, just check configration during init
+    check
+    ;;
   start)
-  	start
-	;;
+    # always check config if running manually
+    [ -z $rcS ] && check
+    start
+    ;;
   stop)
-  	stop
-	;;
+    stop
+    ;;
   restart|reload)
-  	restart
-	;;
+    restart
+    ;;
   *)
-	echo $"Usage: $0 {start|stop|restart}"
-	exit 1
+    echo $"Usage: $0 {start|stop|restart}"
+    exit 1
 esac
 
 exit $?
diff -Naur old/package/openssl/c_rehash new/package/openssl/c_rehash
--- old/package/openssl/c_rehash	1970-01-01 01:00:00.000000000 +0100
+++ new/package/openssl/c_rehash	2012-03-08 21:43:12.468750000 +0000
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+if [ ! $1 ]
+then
+	echo "Usage: $0 <pathname>"
+	exit
+fi
+
+if [ ! -d $1 ]
+then
+	echo "Error: Cannot find directory '$1'"
+	exit
+fi
+
+cd $1
+
+# Remove old hash files
+rm -f [0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z].[0-9]*
+
+# Create hashes of each file, avoiding collisions with an incremental suffix
+ls -1 *.pem | while read cert_file
+do
+	hash=`openssl x509 -subject_hash_old -noout -in $cert_file 2> /dev/null`
+	if [ $? == 0 ]
+	then
+		suffix=0
+		while [ -e $hash.$suffix ]
+		do
+			suffix=`expr $suffix + 1`
+		done
+		ln -s $cert_file $hash.$suffix
+	fi
+done
+
diff -Naur old/package/openssl/openssl.mk new/package/openssl/openssl.mk
--- old/package/openssl/openssl.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/openssl/openssl.mk	2012-07-18 21:57:11.781250000 +0100
@@ -4,7 +4,7 @@
 #
 #############################################################
 
-OPENSSL_VERSION = 1.0.0e
+OPENSSL_VERSION = 1.0.1c
 OPENSSL_SITE = http://www.openssl.org/source
 OPENSSL_INSTALL_STAGING = YES
 OPENSSL_DEPENDENCIES = zlib
@@ -97,6 +97,7 @@
 
 define OPENSSL_INSTALL_FIXUPS
 	rm -f $(TARGET_DIR)/usr/bin/c_rehash
+	cp package/openssl/c_rehash $(TARGET_DIR)/usr/bin/c_rehash
 	# libraries gets installed read only, so strip fails
 	chmod +w $(TARGET_DIR)/usr/lib/engines/lib*.so
 	for i in $(addprefix $(TARGET_DIR)/usr/lib/,libcrypto.so.* libssl.so.*); \
diff -Naur old/package/pppd/pppd-2.4.5-maxconfreq-30000.patch new/package/pppd/pppd-2.4.5-maxconfreq-30000.patch
--- old/package/pppd/pppd-2.4.5-maxconfreq-30000.patch	1970-01-01 01:00:00.000000000 +0100
+++ new/package/pppd/pppd-2.4.5-maxconfreq-30000.patch	2012-09-05 03:38:02.578125000 +0100
@@ -0,0 +1,12 @@
+diff -urN ppp-2.4.5/pppd/fsm.h ppp-2.4.5-lt1/pppd/fsm.h
+--- ppp-2.4.5/pppd/fsm.h	2009-11-16 17:26:07.000000000 -0500
++++ ppp-2.4.5-lt1/pppd/fsm.h	2012-09-04 22:39:37.405603914 -0400
+@@ -145,7 +145,7 @@
+  */
+ #define DEFTIMEOUT	3	/* Timeout time in seconds */
+ #define DEFMAXTERMREQS	2	/* Maximum Terminate-Request transmissions */
+-#define DEFMAXCONFREQS	10	/* Maximum Configure-Request transmissions */
++#define DEFMAXCONFREQS	30000	/* Maximum Configure-Request transmissions */
+ #define DEFMAXNAKLOOPS	5	/* Maximum number of nak loops */
+ 
+ 
diff -Naur old/package/proftpd/proftpd.conf new/package/proftpd/proftpd.conf
--- old/package/proftpd/proftpd.conf	1970-01-01 01:00:00.000000000 +0100
+++ new/package/proftpd/proftpd.conf	2012-11-12 13:25:39.843750000 +0000
@@ -0,0 +1,68 @@
+# This is a basic ProFTPD configuration file (rename it to 
+# 'proftpd.conf' for actual use.  It establishes a single server
+# and a single anonymous login.  It assumes that you have a user/group
+# "nobody" and "ftp" for normal operation and anon.
+
+ServerName			"ProFTPD Default Installation"
+ServerType			standalone
+DefaultServer			on
+UseReverseDNS			off
+IdentLookups			off
+ServerIdent on			"FTP Server Ready."
+
+# Port 21 is the standard FTP port.
+Port				21
+
+# Don't use IPv6 support by default.
+UseIPv6				off
+
+# Umask 022 is a good standard umask to prevent new dirs and files
+# from being group and world writable.
+Umask				022
+
+# To prevent DoS attacks, set the maximum number of child processes
+# to 30.  If you need to allow more than 30 concurrent connections
+# at once, simply increase this value.  Note that this ONLY works
+# in standalone mode, in inetd mode you should use an inetd server
+# that allows you to limit maximum number of processes per service
+# (such as xinetd).
+MaxInstances			30
+
+# Set the user and group under which the server will run.
+User				nobody
+Group				nogroup
+
+# To cause every FTP user to be "jailed" (chrooted) into their home
+# directory, uncomment this line.
+#DefaultRoot ~
+
+# Normally, we want files to be overwriteable.
+AllowOverwrite		on
+
+# Bar use of SITE CHMOD by default
+<Limit SITE_CHMOD>
+  DenyAll
+</Limit>
+
+# A basic anonymous configuration, no upload directories.  If you do not
+# want anonymous users, simply delete this entire <Anonymous> section.
+<Anonymous ~ftp>
+  User				ftp
+  Group				ftp
+
+  # We want clients to be able to login with "anonymous" as well as "ftp"
+  UserAlias			anonymous ftp
+
+  # Limit the maximum number of anonymous logins
+  MaxClients			10
+
+  # We want 'welcome.msg' displayed at login, and '.message' displayed
+  # in each newly chdired directory.
+  DisplayLogin			welcome.msg
+  DisplayChdir			.message
+
+  # Limit WRITE everywhere in the anonymous chroot
+  <Limit WRITE>
+    DenyAll
+  </Limit>
+</Anonymous>
diff -Naur old/package/proftpd/proftpd.mk new/package/proftpd/proftpd.mk
--- old/package/proftpd/proftpd.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/proftpd/proftpd.mk	2011-12-16 02:51:39.281250000 +0000
@@ -33,12 +33,12 @@
 		$(INSTALL) -m 0644 -D $(@D)/sample-configurations/basic.conf $(TARGET_DIR)/etc/proftpd.conf; \
 		$(if $(BR2_INET_IPV6),,$(SED) 's/^UseIPv6/# UseIPv6/' $(TARGET_DIR)/etc/proftpd.conf;) \
 	fi
-	$(INSTALL) -m 0755 package/proftpd/S50proftpd $(TARGET_DIR)/etc/init.d
+	$(INSTALL) -m 0755 package/proftpd/S50proftpd $(TARGET_DIR)/etc/init.d/opt
 endef
 
 define PROFTPD_UNINSTALL_TARGET_CMDS
 	rm -f $(TARGET_DIR)/$(PROFTPD_TARGET_BINARY)
-	rm -f $(TARGET_DIR)/etc/init.d/S50proftpd
+	rm -f $(TARGET_DIR)/etc/init.d/opt/S50proftpd
 	rm -f $(TARGET_DIR)/etc/proftpd.conf
 endef
 
diff -Naur old/package/regulatory/Config.in new/package/regulatory/Config.in
--- old/package/regulatory/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ new/package/regulatory/Config.in	2012-06-28 20:04:06.984375000 +0100
@@ -0,0 +1,7 @@
+config BR2_PACKAGE_REGULATORY
+	bool "regulatory"
+	help
+	  Wireless regulatory database.
+
+	  http://linuxwireless.org/en/developers/Regulatory
+
diff -Naur old/package/regulatory/regulatory.mk new/package/regulatory/regulatory.mk
--- old/package/regulatory/regulatory.mk	1970-01-01 01:00:00.000000000 +0100
+++ new/package/regulatory/regulatory.mk	2012-02-13 04:35:04.562500000 +0000
@@ -0,0 +1,19 @@
+#############################################################
+#
+# regulatory
+#
+#############################################################
+REGULATORY_VERSION:=2011.04.28
+REGULATORY_SOURCE:=wireless-regdb-$(REGULATORY_VERSION).tar.bz2
+REGULATORY_SITE:=http://linuxwireless.org/download/wireless-regdb
+
+define REGULATORY_INSTALL_TARGET_CMDS
+	mkdir -p $(TARGET_DIR)/usr/lib/crda
+	DESTDIR=$(TARGET_DIR) $(MAKE1) -C $(@D) install
+endef
+
+$(eval $(call GENTARGETS))
+
+# ifeq ($(strip $(BR2_PACKAGE_CRDA)),y)
+TARGETS+=regulatory
+# endif
diff -Naur old/package/socat/socat.mk new/package/socat/socat.mk
--- old/package/socat/socat.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/socat/socat.mk	2012-12-05 04:11:16.796875000 +0000
@@ -4,7 +4,7 @@
 #
 #############################################################
 
-SOCAT_VERSION = 2.0.0-b4
+SOCAT_VERSION = 2.0.0-b5
 SOCAT_SOURCE = socat-$(SOCAT_VERSION).tar.bz2
 SOCAT_SITE = http://www.dest-unreach.org/socat/download/
 SOCAT_CONF_ENV = sc_cv_termios_ispeed=no \
diff -Naur old/package/udev/Config.in new/package/udev/Config.in
--- old/package/udev/Config.in	2011-11-30 23:08:44.000000000 +0000
+++ new/package/udev/Config.in	2012-08-12 15:54:38.203125000 +0100
@@ -1,29 +1,33 @@
 config BR2_PACKAGE_UDEV
 	bool "udev"
 	depends on BR2_ROOTFS_DEVICE_CREATION_DYNAMIC_UDEV
+	depends on BR2_LARGEFILE # util-linux
+	depends on BR2_USE_WCHAR # util-linux
+	select BR2_PACKAGE_UTIL_LINUX
+	select BR2_PACKAGE_UTIL_LINUX_LIBBLKID
+	select BR2_PACKAGE_KMOD
 	help
 	  Userspace device daemon.
 
+	  udev requires a Linux kernel >= 2.6.34: it relies on devtmpfs.
+
 	  ftp://ftp.kernel.org/pub/linux/utils/kernel/hotplug/
 
 if BR2_PACKAGE_UDEV
 
+config BR2_PACKAGE_UDEV_RULES_GEN
+	bool "enable rules generator"
+	help
+	  Enable persistant rules generator
+
 config BR2_PACKAGE_UDEV_ALL_EXTRAS
 	bool "enable all extras"
-	depends on BR2_LARGEFILE # acl
-	depends on BR2_USE_WCHAR # libglib2
 	select BR2_PACKAGE_ACL
-	select BR2_PACKAGE_LIBUSB
-	select BR2_PACKAGE_LIBUSB_COMPAT
-	select BR2_PACKAGE_USBUTILS
 	select BR2_PACKAGE_HWDATA
 	select BR2_PACKAGE_LIBGLIB2
 	help
 	  Enable all extras with external dependencies like
-	  libacl, libusb, libusb-compat, usbutils, hwdata
-
-comment "udev extras requires a toolchain with LARGEFILE + WCHAR support"
-	depends on !(BR2_LARGEFILE && BR2_USE_WCHAR)
+	  libacl, hwdata and libglib2
 
 endif
 
diff -Naur old/package/udev/S10udev new/package/udev/S10udev
--- old/package/udev/S10udev	2011-11-30 23:08:44.000000000 +0000
+++ new/package/udev/S10udev	2012-10-22 20:59:22.687500000 +0100
@@ -1,47 +1,56 @@
 #!/bin/sh
 #
-# udev	This is a minimal non-LSB version of a UDEV startup script.  It
-#	was derived by stripping down the udev-058 LSB version for use
-#	with buildroot on embedded hardware using Linux 2.6.12+ kernels.
+#   This is a minimal non-LSB version of a UDEV startup script.  It
+#   was derived by stripping down the udev-058 LSB version for use
+#   with buildroot on embedded hardware using Linux 2.6.34+ kernels.
 #
-#	You may need to customize this for your system's resource limits
-#	(including startup time!) and administration.  For example, if
-#	your early userspace has a custom initramfs or initrd you might
-#	need /dev much earlier; or without hotpluggable busses (like USB,
-#	PCMCIA, MMC/SD, and so on) your /dev might be static after boot.
+#   You may need to customize this for your system's resource limits
+#   (including startup time!) and administration.  For example, if
+#   your early userspace has a custom initramfs or initrd you might
+#   need /dev much earlier; or without hotpluggable busses (like USB,
+#   PCMCIA, MMC/SD, and so on) your /dev might be static after boot.
 #
-#	This script assumes your system boots right into the eventual root
-#	filesystem, and that init runs this udev script before any programs
-#	needing more device nodes than the bare-bones set -- /dev/console,
-#	/dev/zero, /dev/null -- that's needed to boot and run this script.
+#   This script assumes your system boots right into the eventual root
+#   filesystem, and that init runs this udev script before any programs
+#   needing more device nodes than the bare-bones set -- /dev/console,
+#   /dev/zero, /dev/null -- that's needed to boot and run this script.
 #
 
-# Check for missing binaries
-UDEV_BIN=/sbin/udevd
+# check for missing binaries
+UDEV_BIN=/lib/udev/udevd
 test -x $UDEV_BIN || exit 5
 
-# Check for config file and read it
-UDEV_CONFIG=/etc/udev/udev.conf
-test -r $UDEV_CONFIG || exit 6
-. $UDEV_CONFIG
-
-case "$1" in
-    start)
-        echo -n "Populating $udev_root using udev: "
-        echo -e '\000\000\000\000' > /proc/sys/kernel/hotplug
-        $UDEV_BIN -d || (echo "FAIL" && exit 1)
-        echo "done"
-        ;;
-    stop)
-        # Stop execution of events
-        udevadm control --stop_exec_queue
-        killall udevd
-        ;;
-    *)
-        echo "Usage: $0 {start|stop}"
-        exit 1
-        ;;
-esac
+# check for config file and read it
+UDEV_CONF=/etc/udev/udev.conf
+test -r $UDEV_CONF || exit 6
+source $UDEV_CONF
+
+case $1 in
+
+  stop)
+    echo -n "Stopping execution of events, via udev... "
+    udevadm control --stop-exec-queue
+    killall udevd 2>/dev/null
+    echo
+    ;;
 
+  start)
+    echo -n "Starting udev daemon, populating ${udev_root:-/dev} "
+    echo -e '\000\000\000\000' >/proc/sys/kernel/hotplug
+    $UDEV_BIN -d || { echo "ERROR"; exit 1; }
+    udevadm trigger
+    echo
+    ;;
+
+  *)
+    echo "Usage: $0 {stop|start}"
+    exit 1
+    ;;
+esac
 
 exit 0
+##
+## report stuff relative to a device node
+##  udevadm info -a -n </dev/device>
+##
+
diff -Naur old/package/udev/udev.mk new/package/udev/udev.mk
--- old/package/udev/udev.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/udev/udev.mk	2012-08-25 13:37:36.328125000 +0100
@@ -3,35 +3,58 @@
 # udev
 #
 #############################################################
-UDEV_VERSION = 173
+UDEV_VERSION = 182
 UDEV_SOURCE = udev-$(UDEV_VERSION).tar.bz2
 UDEV_SITE = $(BR2_KERNEL_MIRROR)/linux/utils/kernel/hotplug/
 UDEV_INSTALL_STAGING = YES
 
+# mq_getattr is in librt
+UDEV_CONF_ENV += LIBS=-lrt
+
 UDEV_CONF_OPT =			\
 	--sbindir=/sbin		\
 	--with-rootlibdir=/lib	\
-	--libexecdir=/lib/udev	\
+	--libexecdir=/lib	\
+	--with-usb-ids-path=/usr/share/hwdata/usb.ids	\
+	--with-pci-ids-path=/usr/share/hwdata/pci.ids	\
+	--with-firmware-path=/lib/firmware		\
 	--disable-introspection
 
-UDEV_DEPENDENCIES = host-gperf host-pkg-config
+UDEV_DEPENDENCIES = host-gperf host-pkg-config util-linux kmod
+
+ifeq ($(BR2_PACKAGE_UDEV_RULES_GEN),y)
+UDEV_CONF_OPT += --enable-rule_generator
+endif
 
 ifeq ($(BR2_PACKAGE_UDEV_ALL_EXTRAS),y)
-UDEV_DEPENDENCIES += libusb libusb-compat acl usbutils hwdata libglib2
-UDEV_CONF_OPT +=							\
-	--with-pci-ids-path=$(TARGET_DIR)/usr/share/hwdata/pci.ids	\
-	--with-usb-ids-path=$(TARGET_DIR)/usr/share/hwdata/usb.ids	\
+UDEV_DEPENDENCIES += acl hwdata libglib2
+UDEV_CONF_OPT +=		\
 	--enable-udev_acl
 else
 UDEV_CONF_OPT +=		\
-	--disable-hwdb		\
 	--disable-gudev
 endif
 
+ifeq ($(BR2_PACKAGE_SYSTEMD),y)
+	UDEV_CONF_OPT += --with-systemdsystemunitdir=/lib/systemd/system/
+endif
+
 define UDEV_INSTALL_INITSCRIPT
 	$(INSTALL) -m 0755 package/udev/S10udev $(TARGET_DIR)/etc/init.d/S10udev
 endef
 
+define UDEV_NO_AUTO_LOAD_HANDLING
+	@echo UDEV_NO_AUTO_LOAD_HANDLING 
+	# Avoid auto-loading all device drivers, with udev_182. (HACK)
+	# Some devices require firmware and are highly dependent on various conditions
+	# and thus handled by specific scripts.
+	( cd $(TARGET_DIR)/lib/udev; \
+	  [ -f rules.d/??-drivers.rules ] \
+	  && { mkdir -p rules.disabled; mv rules.d/??-drivers.rules rules.disabled; } || : \
+	)
+endef
+
 UDEV_POST_INSTALL_TARGET_HOOKS += UDEV_INSTALL_INITSCRIPT
+UDEV_POST_INSTALL_TARGET_HOOKS += UDEV_NO_AUTO_LOAD_HANDLING
 
 $(eval $(call AUTOTARGETS))
diff -Naur old/package/usbmount/usbmount.mk new/package/usbmount/usbmount.mk
--- old/package/usbmount/usbmount.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/usbmount/usbmount.mk	2011-12-14 03:58:49.265625000 +0000
@@ -16,7 +16,7 @@
 	$(INSTALL) -m 0755 -D $(@D)/00_remove_model_symlink 	\
 		$(TARGET_DIR)/etc/usbmount/usbmount.d/00_remove_model_symlink
 
-	$(INSTALL) -m 0644 -D $(@D)/usbmount.rules $(TARGET_DIR)/lib/udev/rules.d/usbmount.rules
+	$(INSTALL) -m 0644 -D $(@D)/usbmount.rules $(TARGET_DIR)/etc/udev/rules.d/usbmount.rules
 	@if [ ! -f $(TARGET_DIR)/etc/usbmount/usbmount.conf ]; then \
 	        $(INSTALL) -m 0644 -D $(@D)/usbmount.conf $(TARGET_DIR)/etc/usbmount/usbmount.conf; \
 	fi
@@ -27,7 +27,7 @@
 define USBMOUNT_UNINSTALL_TARGET_CMDS
 	rm -rf $(TARGET_DIR)/etc/usbmount			\
 		$(TARGET_DIR)/usr/share/usbmount/usbmount	\
-		$(TARGET_DIR)/lib/udev/rules.d/usbmount.rules	\
+		$(TARGET_DIR)/etc/udev/rules.d/usbmount.rules	\
 		$(TARGET_DIR)/media/usb?
 endef
 
diff -Naur old/package/vpnc/vpnc.mk new/package/vpnc/vpnc.mk
--- old/package/vpnc/vpnc.mk	2011-11-30 23:08:44.000000000 +0000
+++ new/package/vpnc/vpnc.mk	2012-04-13 19:22:51.656250000 +0100
@@ -10,7 +10,7 @@
 VPNC_DIR=$(BUILD_DIR)/vpnc-$(VPNC_VERSION)
 VPNC_CAT:=$(ZCAT)
 VPNC_BINARY:=$(VPNC_DIR)/vpnc
-VPNC_DEST_DIR:=$(TARGET_DIR)/usr/local/sbin
+VPNC_DEST_DIR:=$(TARGET_DIR)/usr/sbin
 VPNC_TARGET_BINARY:=$(VPNC_DEST_DIR)/vpnc
 VPNC_TARGET_SCRIPT:=$(TARGET_DIR)/etc/vpnc/default.conf
 
@@ -32,8 +32,8 @@
 $(VPNC_TARGET_BINARY): $(VPNC_BINARY)
 	$(MAKE) $(TARGET_CONFIGURE_OPTS) \
 		DESTDIR=$(TARGET_DIR) \
-		BINDIR=/usr/local/bin \
-		SBINDIR=/usr/local/sbin \
+		BINDIR=/usr/bin \
+		SBINDIR=/usr/sbin \
 		ETCDIR=/etc/vpnc \
 		MANDIR=/usr/share/man \
 		VERSION=$(VPNC_VERSION) \
diff -Naur old/boot/uboot/uboot.mk new/boot/uboot/uboot.mk
--- old/boot/uboot/uboot.mk	2011-11-30 18:08:44.000000000 -0500
+++ new/boot/uboot/uboot.mk	2013-02-14 01:07:57.706937117 -0500
@@ -82,10 +82,17 @@
 	$(TARGET_CONFIGURE_OPTS) $(UBOOT_CONFIGURE_OPTS) 	\
 		$(MAKE) -C $(@D) $(UBOOT_MAKE_OPTS) 		\
 		$(UBOOT_MAKE_TARGET)
+	# build env tool from same source
+	$(MAKE) -C $(@D) $(UBOOT_MAKE_OPTS) \
+		HOSTCC=$(TARGET_CROSS)gcc HOSTSTRIP=$(TARGET_CROSS)strip env
 endef
 
 define UBOOT_INSTALL_IMAGES_CMDS
 	cp -dpf $(@D)/$(UBOOT_BIN) $(BINARIES_DIR)/
+	# install env tool
+	install -m 0644 -D $(@D)/tools/env/fw_env.config $(TARGET_DIR)/etc
+	install -m 0755 -D $(@D)/tools/env/fw_printenv $(TARGET_DIR)/usr/sbin
+	ln -sf fw_printenv $(TARGET_DIR)/usr/sbin/fw_setenv
 endef
 
 $(eval $(call GENTARGETS))
@@ -98,3 +105,5 @@
 endif
 endif
 endif
+
+
diff -Naur old/target/generic/device_table_dev.txt new/target/generic/device_table_dev.txt
--- old/target/generic/device_table_dev.txt	2011-11-30 23:08:44.000000000 +0000
+++ new/target/generic/device_table_dev.txt	2011-12-14 03:58:49.265625000 +0000
@@ -46,8 +46,8 @@
 
 
 # MTD stuff
-/dev/mtd	c	640	0	0	90	0	0	2	4
-/dev/mtdblock	b	640	0	0	31	0	0	1	4
+/dev/mtd	c	640	0	0	90	0	0	2	8
+/dev/mtdblock	b	640	0	0	31	0	0	1	8
 
 #Tun/tap driver
 /dev/net	d	755	0	0	-	-	-	-	-
diff -Naur old/target/generic/device_table.txt new/target/generic/device_table.txt
--- old/target/generic/device_table.txt	2011-11-30 23:08:44.000000000 +0000
+++ new/target/generic/device_table.txt	2012-05-25 22:07:28.531250000 +0100
@@ -10,6 +10,7 @@
 /tmp					d	1777	0	0	-	-	-	-	-
 /etc					d	755	0	0	-	-	-	-	-
 /home/default				d	2755	1000	1000	-	-	-	-	-
+/home/summit				d	2755	1001	1001	-	-	-	-	-
 /bin/busybox				f	4755	0	0	-	-	-	-	-
 /etc/shadow				f	600	0	0	-	-	-	-	-
 /etc/passwd				f	644	0	0	-	-	-	-	-
-/etc/network/if-up.d         d  755   0  0  -  -  -  -  -
-/etc/network/if-pre-up.d     d  755   0  0  -  -  -  -  -
-/etc/network/if-down.d       d  755   0  0  -  -  -  -  -
-/etc/network/if-post-down.d     d  755   0  0  -  -  -  -  -
-/usr/share/udhcpc/default.script   f  755   0  0  -  -  -  -  -
-# uncomment this to allow starting x as non-root
-#/usr/X11R6/bin/Xfbdev          f  4755  0  0  -  -  -  -  -
