diff -Naur old/boot/at91bootstrap3/at91bootstrap3.mk new/boot/at91bootstrap3/at91bootstrap3.mk
--- old/boot/at91bootstrap3/at91bootstrap3.mk	2013-04-22 23:49:35.189440176 +0100
+++ new/boot/at91bootstrap3/at91bootstrap3.mk	2013-04-29 10:06:40.000000000 +0100
@@ -3,10 +3,10 @@
 # at91bootstrap3
 #
 #############################################################
-AT91BOOTSTRAP3_VERSION = 3.2
+AT91BOOTSTRAP3_VERSION = 3.4.4
 AT91BOOTSTRAP3_SITE = \
 	ftp://www.at91.com/pub/at91bootstrap/AT91Bootstrap$(AT91BOOTSTRAP3_VERSION)
-AT91BOOTSTRAP3_SOURCE = at91bootstrap_9n12.tar.gz
+AT91BOOTSTRAP3_SOURCE = AT91Bootstrap-v3.4.4.tar.xz
 
 AT91BOOTSTRAP3_INSTALL_IMAGES = YES
 AT91BOOTSTRAP3_INSTALL_TARGET = NO
@@ -21,11 +21,14 @@
 AT91BOOTSTRAP3_MAKE_OPT = CROSS_COMPILE=$(TARGET_CROSS) DESTDIR=$(BINARIES_DIR)
 
 ifneq ($(AT91BOOTSTRAP3_CUSTOM_PATCH_DIR),)
+define AT91BOOTSTRAP3_REMOVE_SPECIFIC_PATCHES
+	rm -fv boot/at91bootstrap3/at91bootstrap3-u-boot-relocation-fix.patch
+endef
 define AT91BOOTSTRAP3_APPLY_CUSTOM_PATCHES
 	support/scripts/apply-patches.sh $(@D) $(AT91BOOTSTRAP3_CUSTOM_PATCH_DIR) \
 		at91bootstrap3-\*.patch
 endef
-
+AT91BOOTSTRAP3_PRE_PATCH_HOOKS += AT91BOOTSTRAP3_REMOVE_SPECIFIC_PATCHES
 AT91BOOTSTRAP3_POST_PATCH_HOOKS += AT91BOOTSTRAP3_APPLY_CUSTOM_PATCHES
 endif
 
@@ -42,11 +45,17 @@
 endef
 
 define AT91BOOTSTRAP3_BUILD_CMDS
-	$(MAKE) $(AT91BOOTSTRAP3_MAKE_OPT) -C $(@D) boot
+	$(MAKE) $(AT91BOOTSTRAP3_MAKE_OPT) -C $(@D)
 endef
 
+AT91BOOTSTRAP3_BIN_FILE = $(call qstrip,$(BR2_SDC_PLATFORM))-nandflashboot-uboot.bin
+ifeq ($(call qstrip,$(BR2_SDC_PLATFORM)),wb45n)
+AT91BOOTSTRAP3_BIN_FILE = at91sam9x5ek-nandflashboot-uboot.bin
+endif
+
 define AT91BOOTSTRAP3_INSTALL_IMAGES_CMDS
-	$(MAKE) $(AT91BOOTSTRAP3_MAKE_OPT) -C $(@D) bootstrap
+	$(INSTALL) -m 644 $(@D)/binaries/$(AT91BOOTSTRAP3_BIN_FILE) \
+	                  $(BINARIES_DIR)/$(call qstrip,$(BR2_SDC_PLATFORM)).bin
 endef
 
 $(eval $(generic-package))
diff -Naur old/boot/uboot/uboot.mk new/boot/uboot/uboot.mk
--- old/boot/uboot/uboot.mk	2013-04-22 23:49:35.197440176 +0100
+++ new/boot/uboot/uboot.mk	2013-04-22 21:36:40.000000000 +0100
@@ -93,6 +93,9 @@
 	$(TARGET_CONFIGURE_OPTS) $(UBOOT_CONFIGURE_OPTS) 	\
 		$(MAKE) -C $(@D) $(UBOOT_MAKE_OPTS) 		\
 		$(UBOOT_MAKE_TARGET)
+	# build env tool from same source
+	$(MAKE) -C $(@D) $(UBOOT_MAKE_OPTS) \
+		HOSTCC=$(TARGET_CROSS)gcc HOSTSTRIP=$(TARGET_CROSS)strip env
 endef
 
 define UBOOT_BUILD_OMAP_IFT
@@ -102,6 +105,10 @@
 
 define UBOOT_INSTALL_IMAGES_CMDS
 	cp -dpf $(@D)/$(UBOOT_BIN) $(BINARIES_DIR)/
+	# install env tool
+	install -m 0644 -D $(@D)/tools/env/fw_env.config $(TARGET_DIR)/etc
+	install -m 0755 -D $(@D)/tools/env/fw_printenv $(TARGET_DIR)/usr/sbin
+	ln -sf fw_printenv $(TARGET_DIR)/usr/sbin/fw_setenv
 	$(if $(BR2_TARGET_UBOOT_SPL),
 		cp -dpf $(@D)/$(BR2_TARGET_UBOOT_SPL_NAME) $(BINARIES_DIR)/)
 endef
diff -Naur old/Config.in new/Config.in
--- old/Config.in	2013-04-22 23:49:34.925440166 +0100
+++ new/Config.in	2013-04-22 21:36:40.000000000 +0100
@@ -454,3 +454,5 @@
 source "linux/Config.in"
 
 source "Config.in.legacy"
+
+source "board/sdc/Config.in"
diff -Naur old/linux/linux.mk new/linux/linux.mk
--- old/linux/linux.mk	2013-04-22 23:49:35.197440176 +0100
+++ new/linux/linux.mk	2013-04-22 21:36:40.000000000 +0100
@@ -139,7 +139,7 @@
 		if echo $$p | grep -q -E "^ftp://|^http://" ; then \
 			support/scripts/apply-patches.sh $(@D) $(DL_DIR) `basename $$p` ; \
 		elif test -d $$p ; then \
-			support/scripts/apply-patches.sh $(@D) $$p linux-\*.patch ; \
+			support/scripts/apply-patches.sh $(@D) $$p linux-$(LINUX_VERSION)\*.patch ; \
 		else \
 			support/scripts/apply-patches.sh $(@D) `dirname $$p` `basename $$p` ; \
 		fi \
diff -Naur old/package/busybox/Config.in new/package/busybox/Config.in
--- old/package/busybox/Config.in	2013-07-18 21:39:00.000000000 -0500
+++ new/package/busybox/Config.in 	2013-07-18 21:38:43.000000000 -0500
@@ -32,7 +32,7 @@
 endchoice
 
 config BR2_BUSYBOX_VERSION
-	string
+	string "BusyBox version to use?"
 	default "1.19.4"	if BR2_BUSYBOX_VERSION_1_19_X
 	default "1.20.2"	if BR2_BUSYBOX_VERSION_1_20_X
 	default "1.21.0"	if BR2_BUSYBOX_VERSION_1_21_X
diff -Naur old/package/busybox/busybox-networking-ifplugd-iface-err.patch new/package/busybox/busybox-networking-ifplugd-iface-err.patch
--- old/package/busybox/busybox-networking-ifplugd-iface-err.patch	1970-01-01 01:00:00.000000000 +0100
+++ new/package/busybox/busybox-networking-ifplugd-iface-err.patch	2013-04-25 21:36:40.000000000 +0100
@@ -0,0 +1,25 @@
+--- busybox-1.21.0/networking/ifplugd.c	2013-04-25 21:35:01.543521939 -0500
++++ busybox-1.21.0/networking/ifplugd.c	2013-04-25 21:35:21.543521939 -0500
+@@ -685,6 +685,11 @@
+ 						G.iface_exists ? "" : "dis");
+ 				if (G.iface_exists)
+ 					maybe_up_new_iface();
++				else {
++					/* notify upon status:  error -> down */
++					iface_status_old = IFSTATUS_ERR;
++					delay_time = 0;
++				}
+ 			}
+ 		}
+ 
+@@ -693,6 +698,7 @@
+ 		if (iface_status == IFSTATUS_ERR) {
+ 			if (!(opts & FLAG_MONITOR))
+ 				goto exiting;
++
+ 			iface_status = IFSTATUS_DOWN;
+ 		}
+ 		iface_status_str = strstatus(iface_status);
+;
+; Allow script notification of a disappeared iface, so can take action.
+;
diff -Naur old/package/busybox/busybox-networking-ifplugd-nl-buf8k.patch new/package/busybox/busybox-networking-ifplugd-nl-buf8k.patch
--- old/package/busybox/busybox-networking-ifplugd-nl-buf8k.patch	1969-12-31 18:00:00.000000000 -0600
+++ new/package/busybox/busybox-networking-ifplugd-nl-buf8k.patch	2013-09-27 13:51:27.726892193 -0500
@@ -0,0 +1,40 @@
+--- busybox-1.21.1/networking/ifplugd.c	2013-06-29 09:58:06.000000000 -0500
++++ busybox-1.21.1/networking/ifplugd.c	2013-09-27 12:37:11.000000000 -0500
+@@ -451,17 +451,25 @@
+ static NOINLINE int check_existence_through_netlink(void)
+ {
+ 	int iface_len;
+-	char replybuf[1024];
++	/*
++	 * Messages in/aft linux-3.8 can exceed 1K buffer size, so increased to 8K.
++	 * netlink.h: "limit to 8K to avoid MSG_TRUNC when PAGE_SIZE is very large".
++	 *            int BUF_SIZE = MIN(sysconf(_SC_PAGESIZE), 8192);
++	 * Note:
++	 * No need to free replybuf on recv error.
++	 */
++	enum { BUF_SIZE = 8 * 1024 };
++	char *replybuf = xmalloc(BUF_SIZE);
+ 
+ 	iface_len = strlen(G.iface);
+ 	while (1) {
+ 		struct nlmsghdr *mhdr;
+ 		ssize_t bytes;
+ 
+-		bytes = recv(netlink_fd, &replybuf, sizeof(replybuf), MSG_DONTWAIT);
++		bytes = recv(netlink_fd, replybuf, BUF_SIZE, MSG_DONTWAIT);
+ 		if (bytes < 0) {
+ 			if (errno == EAGAIN)
+-				return G.iface_exists;
++				goto ret;
+ 			if (errno == EINTR)
+ 				continue;
+ 
+@@ -507,6 +515,8 @@
+ 		}
+ 	}
+ 
++ ret:
++	free(replybuf);
+ 	return G.iface_exists;
+ }
+
diff --git old/package/busybox/busybox-networking-udhcp-new-script-states.patch new/package/busybox/busybox-networking-udhcp-new-script-states.patch
--- old/dev/null
+++ new/package/busybox/busybox-networking-udhcp-new-script-states.patch
@@ -0,0 +1,43 @@
+--- busybox-1.21.1/networking/udhcp/dhcpc.c
++++ busybox-1.21.1/networking/udhcp/dhcpc.c
+@@ -1074,6 +1074,7 @@ static void perform_renew(void)
+ 	case RENEWING:
+ 	case REBINDING:
+ 		state = RENEW_REQUESTED;
++		udhcp_run_script(NULL, "renewing");
+ 		break;
+ 	case RENEW_REQUESTED: /* impatient are we? fine, square 1 */
+ 		udhcp_run_script(NULL, "deconfig");
+@@ -1106,6 +1107,7 @@ static void perform_release(uint32_t server_addr, uint32_t requested_ip)
+ 
+ 	change_listen_mode(LISTEN_NONE);
+ 	state = RELEASED;
++	udhcp_run_script(NULL, "released");
+ }
+ 
+ static uint8_t* alloc_dhcp_option(int code, const char *str, int extra)
+@@ -1488,6 +1490,7 @@ int udhcpc_main(int argc UNUSED_PARAM, char **argv)
+ 			case BOUND:
+ 				/* 1/2 lease passed, enter renewing state */
+ 				state = RENEWING;
++				udhcp_run_script(NULL, "renewing");
+ 				client_config.first_secs = 0; /* make secs field count from 0 */
+ 				change_listen_mode(LISTEN_KERNEL);
+ 				log1("Entering renew state");
+@@ -1512,6 +1515,7 @@ int udhcpc_main(int argc UNUSED_PARAM, char **argv)
+ 				/* Timed out, enter rebinding state */
+ 				log1("Entering rebinding state");
+ 				state = REBINDING;
++				udhcp_run_script(NULL, "rebinding");
+ 				/* fall right through */
+ 			case REBINDING:
+ 				/* Switch to bcast receive */
+@@ -1660,6 +1664,7 @@ int udhcpc_main(int argc UNUSED_PARAM, char **argv)
+ 
+ 				/* enter requesting state */
+ 				state = REQUESTING;
++				udhcp_run_script(NULL, "requesting");
+ 				timeout = 0;
+ 				packet_num = 0;
+ 				already_waited_sec = 0;
+
diff -Naur old/package/Config.in new/package/Config.in
--- old/package/Config.in	2013-04-22 23:49:35.121440173 +0100
+++ new/package/Config.in	2013-04-22 21:36:40.000000000 +0100
@@ -623,7 +623,8 @@
 source "package/can-utils/Config.in"
 source "package/connman/Config.in"
 source "package/ctorrent/Config.in"
+source "package/cifs-utils/Config.in"
 source "package/conntrack-tools/Config.in"
 source "package/cups/Config.in"
 if BR2_PACKAGE_BUSYBOX_SHOW_OTHERS
 source "package/dhcp/Config.in"
diff -Naur old/package/dhcp/dhcp.mk new/package/dhcp/dhcp.mk
--- old/package/dhcp/dhcp.mk	2013-04-22 23:49:35.121440173 +0100
+++ new/package/dhcp/dhcp.mk	2013-04-22 21:36:40.000000000 +0100
@@ -27,7 +27,7 @@
 	(cd $(TARGET_DIR)/var/lib; ln -snf /tmp dhcp)
 	$(INSTALL) -m 0755 -D $(@D)/server/dhcpd $(TARGET_DIR)/usr/sbin/dhcpd
 	$(INSTALL) -m 0755 -D package/dhcp/S80dhcp-server \
-		$(TARGET_DIR)/etc/init.d/S80dhcp-server
+		$(TARGET_DIR)/etc/init.d/opt/S80dhcp-server
 	$(INSTALL) -m 0644 -D package/dhcp/dhcpd.conf \
 		$(TARGET_DIR)/etc/dhcp/dhcpd.conf
 endef
@@ -40,7 +40,7 @@
 	$(INSTALL) -m 0755 -D $(DHCP_DIR)/relay/dhcrelay \
 		$(TARGET_DIR)/usr/sbin/dhcrelay
 	$(INSTALL) -m 0755 -D package/dhcp/S80dhcp-relay \
-		$(TARGET_DIR)/etc/init.d/S80dhcp-relay
+		$(TARGET_DIR)/etc/init.d/opt/S80dhcp-relay
 endef
 endif
 
diff -Naur old/package/ebtables/ebtables-add-arpnat-target.patch new/package/ebtables/ebtables-add-arpnat-target.patch
--- old/package/ebtables/ebtables-add-arpnat-target.patch	1970-01-01 01:00:00.000000000 +0100
+++ new/package/ebtables/ebtables-add-arpnat-target.patch	2013-04-22 21:36:40.000000000 +0100
@@ -0,0 +1,132 @@
+diff -urN ebtables-v2.0.10-2/Makefile ebtables-v2.0.10-2-arpnat/Makefile
+--- ebtables-v2.0.10-2/Makefile	2011-08-11 14:56:16.000000000 -0400
++++ ebtables-v2.0.10-2-arpnat/Makefile	2011-09-18 18:49:20.071816739 -0400
+@@ -17,9 +17,9 @@
+ SYSCONFIGDIR:=/etc/sysconfig
+ DESTDIR:=
+ 
+-CFLAGS:=-Wall -Wunused -Werror
++#CFLAGS:=-Wall -Wunused -Werror
+ CFLAGS_SH_LIB:=-fPIC -O3
+-CC:=gcc
++#CC:=gcc
+ 
+ ifeq ($(shell uname -m),sparc64)
+ CFLAGS+=-DEBT_MIN_ALIGN=8 -DKERNEL_64_USERSPACE_32
+diff -urN ebtables-v2.0.10-2/extensions/ebt_nat.c ebtables-v2.0.10-2-arpnat/extensions/ebt_nat.c
+--- ebtables-v2.0.10-2/extensions/ebt_nat.c	2011-08-11 14:56:16.000000000 -0400
++++ ebtables-v2.0.10-2-arpnat/extensions/ebt_nat.c	2011-09-17 12:59:35.497511502 -0400
+@@ -21,6 +21,7 @@
+ #define NAT_S_TARGET '2'
+ #define NAT_D_TARGET '2'
+ #define NAT_S_ARP '3'
++#define NAT_ARP_TARGET '2'
+ static struct option opts_s[] =
+ {
+ 	{ "to-source"     , required_argument, 0, NAT_S },
+@@ -30,6 +31,12 @@
+ 	{ 0 }
+ };
+ 
++static struct option opts_arpnat[] =
++{
++	{ "arpnat-target" , required_argument, 0, NAT_ARP_TARGET },
++	{ 0 }
++};
++
+ static struct option opts_d[] =
+ {
+ 	{ "to-destination", required_argument, 0, NAT_D },
+@@ -47,6 +54,13 @@
+ 	" --snat-arp             : also change src address in arp msg\n");
+ }
+ 
++static void print_help_arpnat()
++{
++	printf(
++	"arpnat options:\n"
++	" --arpnat-target target   : ACCEPT, DROP, RETURN or CONTINUE\n");
++}
++
+ static void print_help_d()
+ {
+ 	printf(
+@@ -55,6 +69,54 @@
+ 	" --dnat-target target   : ACCEPT, DROP, RETURN or CONTINUE\n");
+ }
+ 
++static void init_arpnat(struct ebt_entry_target *target)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
++
++   natinfo->target = EBT_ACCEPT;
++   return;
++}
++#define OPT_ARPNAT_TARGET 0x2
++static int parse_arpnat(int c, char **argv, int argc,
++   const struct ebt_u_entry *entry, unsigned int *flags,
++   struct ebt_entry_target **target)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)(*target)->data;
++
++   switch (c) {
++   case NAT_ARP_TARGET:
++       ebt_check_option2(flags, OPT_ARPNAT_TARGET);
++       if (FILL_TARGET(optarg, natinfo->target))
++           ebt_print_error2("Illegal --arpnat-target target");
++       break;
++   default:
++       return 0;
++   }
++   return 1;
++}
++
++static void final_check_arpnat(const struct ebt_u_entry *entry,
++   const struct ebt_entry_target *target, const char *name,
++   unsigned int hookmask, unsigned int time)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
++
++   if (BASE_CHAIN && natinfo->target == EBT_RETURN)
++       ebt_print_error("--arpnat-target RETURN not allowed on base chain");
++   CLEAR_BASE_CHAIN_BIT;
++   if (((hookmask & ~((1 << NF_BR_PRE_ROUTING) | (1 << NF_BR_POST_ROUTING)))
++      || strcmp(name, "nat")))
++       ebt_print_error("Wrong chain for arpnat");
++}
++
++static void print_arpnat(const struct ebt_u_entry *entry,
++   const struct ebt_entry_target *target)
++{
++   struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
++
++   printf(" --arpnat-target %s", TARGET_NAME(natinfo->target));
++}
++
+ static void init_s(struct ebt_entry_target *target)
+ {
+ 	struct ebt_nat_info *natinfo = (struct ebt_nat_info *)target->data;
+@@ -231,8 +293,23 @@
+ 	.extra_ops	= opts_d,
+ };
+ 
++#define EBT_ARPNAT_TARGET "arpnat"
++static struct ebt_u_target arpnat_target =
++{
++   .name       = EBT_ARPNAT_TARGET,
++   .size       = sizeof(struct ebt_nat_info),
++   .help       = print_help_arpnat,
++   .init       = init_arpnat,
++   .parse      = parse_arpnat,
++   .final_check    = final_check_arpnat,
++   .print      = print_arpnat,
++   .compare    = compare,
++   .extra_ops  = opts_arpnat,
++};
++
+ void _init(void)
+ {
+ 	ebt_register_target(&snat_target);
+ 	ebt_register_target(&dnat_target);
++	ebt_register_target(&arpnat_target);
+ }
diff -Naur old/package/input-event-daemon/input-event-daemon.mk new/package/input-event-daemon/input-event-daemon.mk
--- old/package/input-event-daemon/input-event-daemon.mk	2013-04-22 23:49:34.961440167 +0100
+++ new/package/input-event-daemon/input-event-daemon.mk	2013-04-22 21:36:40.000000000 +0100
@@ -21,9 +21,9 @@
 	[ -f $(TARGET_DIR)/etc/input-event-daemon.conf ] || \
 		$(INSTALL) -m 644 -D $(@D)/docs/sample.conf \
 			$(TARGET_DIR)/etc/input-event-daemon.conf
-	[ -f $(TARGET_DIR)/etc/init.d/S99input-event-daemon ] || \
+	[ -f $(TARGET_DIR)/etc/init.d/opt/S99input-event-daemon ] || \
 		$(INSTALL) -m 0755 -D package/input-event-daemon/S99input-event-daemon \
-			$(TARGET_DIR)/etc/init.d/S99input-event-daemon
+			$(TARGET_DIR)/etc/init.d/opt/S99input-event-daemon
 endef
 
 define INPUT_EVENT_DAEMON_CLEAN_CMDS
diff -Naur old/package/libnl/libnl.mk new/package/libnl/libnl.mk
--- old/package/libnl/libnl.mk	2013-04-22 23:49:35.189440176 +0100
+++ new/package/libnl/libnl.mk	2013-04-22 21:36:40.000000000 +0100
@@ -4,7 +4,7 @@
 #
 #############################################################
 
-LIBNL_VERSION = 3.2.18
+LIBNL_VERSION = 3.2.21
 LIBNL_SITE = http://www.infradead.org/~tgr/libnl/files
 LIBNL_LICENSE = LGPLv2.1+
 LIBNL_LICENSE_FILES = COPYING
diff -Naur old/package/openssh/openssh.mk new/package/openssh/openssh.mk
--- old/package/openssh/openssh.mk	2013-04-22 23:49:34.941440166 +0100
+++ new/package/openssh/openssh.mk	2013-04-22 21:36:40.000000000 +0100
@@ -8,12 +8,13 @@
 OPENSSH_SITE = http://ftp.openbsd.org/pub/OpenBSD/OpenSSH/portable
 OPENSSH_CONF_ENV = LD="$(TARGET_CC)" LDFLAGS="$(TARGET_CFLAGS)"
 OPENSSH_CONF_OPT = --libexecdir=/usr/lib --disable-lastlog --disable-utmp \
-		--disable-utmpx --disable-wtmp --disable-wtmpx --disable-strip
+		--disable-utmpx --disable-wtmp --disable-wtmpx --disable-strip \
+		--sysconfdir=/etc/ssh
 
 OPENSSH_DEPENDENCIES = zlib openssl
 
 define OPENSSH_INSTALL_INITSCRIPT
-	$(INSTALL) -D -m 755 package/openssh/S50sshd $(TARGET_DIR)/etc/init.d/S50sshd
+	$(INSTALL) -D -m 755 package/openssh/S50sshd $(TARGET_DIR)/etc/init.d/opt/S50sshd
 endef
 
 OPENSSH_POST_INSTALL_TARGET_HOOKS += OPENSSH_INSTALL_INITSCRIPT
diff -Naur old/package/openssh/S50sshd new/package/openssh/S50sshd
--- old/package/openssh/S50sshd	2013-04-22 23:49:34.941440166 +0100
+++ new/package/openssh/S50sshd	2013-09-17 07:07:08.000000000 -0500
@@ -1,72 +1,95 @@
 #!/bin/sh
-#
-# sshd        Starts sshd.
-#
-
-# Make sure the ssh-keygen progam exists
-[ -f /usr/bin/ssh-keygen ] || exit 0
-
-# Check for the SSH1 RSA key
-if [ ! -f /etc/ssh_host_key ] ; then
-	echo Generating RSA Key...
-	/usr/bin/ssh-keygen -t rsa1 -f /etc/ssh_host_key -C '' -N ''
-fi
-
-# Check for the SSH2 RSA key
-if [ ! -f /etc/ssh_host_rsa_key ] ; then
-	echo Generating RSA Key...
-	/usr/bin/ssh-keygen -t rsa -f /etc/ssh_host_rsa_key -C '' -N ''
-fi
-
-# Check for the SSH2 DSA key
-if [ ! -f /etc/ssh_host_dsa_key ] ; then
-	echo Generating DSA Key...
-	echo THIS CAN TAKE A MINUTE OR TWO DEPENDING ON YOUR PROCESSOR!
-	echo
-        /usr/bin/ssh-keygen -t dsa -f /etc/ssh_host_dsa_key -C '' -N ''
-fi
-
-# Check for the SSH2 ECDSA key
-if [ ! -f /etc/ssh_host_ecdsa_key ]; then
-	echo Generating ECDSA Key...
-	echo THIS CAN TAKE A MINUTE OR TWO DEPENDING ON YOUR PROCESSOR!
-	echo
-	/usr/bin/ssh-keygen -t ecdsa -f /etc/ssh_host_ecdsa_key -C '' -N ''
-fi
-                
-umask 077
+# sshd init-script
+# supports inetd and pre-checking configuration
+
+check_configuration() {
+  echo "Checking sshd configuration..."
+  
+  # Make sure the ssh-keygen progam exists
+  [ -x /usr/bin/ssh-keygen ] || { echo "  ...missing ssh-keygen"; exit 1; }
+
+  # Check that the sshd_config file exists
+  [ -f /etc/ssh/sshd_config ] || { echo "  ...missing sshd_config"; exit 1; }
+
+  # See if need to generate the SSH1 RSA key
+  if [ ! -s /etc/ssh/ssh_host_key ] \
+  && grep -q "^HostKey.*host_key" /etc/ssh/sshd_config
+  then
+    echo Generating RSA Key...
+    /usr/bin/ssh-keygen -t rsa1 -f /etc/ssh/ssh_host_key~ -C '' -N '' \
+    && {
+      mv -f /etc/ssh/ssh_host_key~.pub /etc/ssh/ssh_host_key.pub
+      mv -f /etc/ssh/ssh_host_key~ /etc/ssh/ssh_host_key
+    }
+  fi
+
+  # See if need to generate the SSH2 RSA key
+  if [ ! -s /etc/ssh/ssh_host_rsa_key ] \
+  && grep -q "^HostKey.*host_rsa_key" /etc/ssh/sshd_config
+  then
+    echo Generating RSA Key...
+    /usr/bin/ssh-keygen -t rsa -f /etc/ssh/ssh_host_rsa_key~ -C '' -N '' \
+    && {
+      mv -f /etc/ssh/ssh_host_rsa_key~.pub /etc/ssh/ssh_host_rsa_key.pub
+      mv -f /etc/ssh/ssh_host_rsa_key~ /etc/ssh/ssh_host_rsa_key
+    }
+  fi
+
+  # See if need to generate the SSH2 DSA key
+  if [ ! -s /etc/ssh/ssh_host_dsa_key ] \
+  && grep -q "^HostKey.*host_dsa_key" /etc/ssh/sshd_config
+  then
+    echo Generating DSA Key... \(may take awhile\)
+    /usr/bin/ssh-keygen -t dsa -f /etc/ssh/ssh_host_dsa_key~ -C '' -N '' \
+    && {
+      mv -f /etc/ssh/ssh_host_dsa_key~.pub /etc/ssh/ssh_host_dsa_key.pub
+      mv -f /etc/ssh/ssh_host_dsa_key~ /etc/ssh/ssh_host_dsa_key
+    }
+  fi
+
+  # See if need to generate the SSH2 ECDSA key
+  if [ ! -s /etc/ssh/ssh_host_ecdsa_key ] \
+  && grep -q "^HostKey.*host_ecdsa_key" /etc/ssh/sshd_config
+  then
+    echo Generating ECDSA Key... \(may take awhile\)
+    /usr/bin/ssh-keygen -t ecdsa -f /etc/ssh/ssh_host_ecdsa_key~ -C '' -N '' \
+    && {
+      mv -f /etc/ssh/ssh_host_ecdsa_key~.pub /etc/ssh/ssh_host_ecdsa_key.pub
+      mv -f /etc/ssh/ssh_host_ecdsa_key~ /etc/ssh/ssh_host_ecdsa_key
+    }
+  fi
+}                
 
-start() {
- 	echo -n "Starting sshd: "
-	/usr/sbin/sshd
-	touch /var/lock/sshd
-	echo "OK"
-}	
-stop() {
-	echo -n "Stopping sshd: "
-        killall	sshd 
-	rm -f /var/lock/sshd
-	echo "OK" 
-}
-restart() {
-	stop
-	start
-}	
+umask 077
 
 case "$1" in
-  start)
-  	start
-	;;
+  check)
+    # inetd pre-check support
+    # Ror each enabled key type in config, the respective keyfile is checked to
+    # exist and be greater than zero size. The key pair are regenerated w/temp
+    # names and upon completion, renamed to respective names.
+    check_configuration
+    ;;
   stop)
-  	stop
-	;;
-  restart|reload)
-  	restart
-	;;
+    echo -n "Stopping sshd: "
+    killall sshd 2>/dev/null
+    #rm -f /var/lock/sshd.lock
+    echo "OK" 
+    ;;
+  start)
+    # always check config if running manually
+    [ -z "$rcS_" ] && check_configuration
+    echo -n "Starting sshd: "
+    /usr/sbin/sshd
+    #touch /var/lock/sshd.lock
+    echo "OK"
+    ;;
+  restart)
+    $0 stop
+    $0 start
+    ;;
   *)
-	echo "Usage: $0 {start|stop|restart}"
-	exit 1
+    echo "Usage: $0 {stop|start|restart}"
+    false
 esac
-
 exit $?
-
diff -Naur old/package/openssl/c_rehash new/package/openssl/c_rehash
--- old/package/openssl/c_rehash	1970-01-01 01:00:00.000000000 +0100
+++ new/package/openssl/c_rehash	2013-04-22 21:36:40.000000000 +0100
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+if [ ! $1 ]
+then
+	echo "Usage: $0 <pathname>"
+	exit
+fi
+
+if [ ! -d $1 ]
+then
+	echo "Error: Cannot find directory '$1'"
+	exit
+fi
+
+cd $1
+
+# Remove old hash files
+rm -f [0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z].[0-9]*
+
+# Create hashes of each file, avoiding collisions with an incremental suffix
+ls -1 *.pem | while read cert_file
+do
+	hash=`openssl x509 -subject_hash_old -noout -in $cert_file 2> /dev/null`
+	if [ $? == 0 ]
+	then
+		suffix=0
+		while [ -e $hash.$suffix ]
+		do
+			suffix=`expr $suffix + 1`
+		done
+		ln -s $cert_file $hash.$suffix
+	fi
+done
+
diff -Naur old/package/openssl/openssl.mk new/package/openssl/openssl.mk
--- old/package/openssl/openssl.mk	2013-04-28 21:37:01.207738342 -0500
+++ new/package/openssl/openssl.mk	2013-05-06 08:23:23.210851615 -0500
@@ -13,6 +13,15 @@
 OPENSSL_TARGET_ARCH = generic32
 OPENSSL_CFLAGS = $(TARGET_CFLAGS)
 
+# require openssl-fips built firstly
+ifneq ($(BR2_PACKAGE_OPENSSL_FIPS),)
+OPENSSL_DEPENDENCIES += openssl-fips
+OPENSSL_FIPS_CFG = fips
+OPENSSL_FIPS_OPT = FIPSDIR=$(STAGING_DIR)/usr/local/ssl/fips-2.0 \
+                   FIPS_SIG=$(STAGING_DIR)/usr/local/ssl/fips-2.0/bin/incore
+OPENSSL_FIPS_MAKE_OPT = FIPS_SIG=$(STAGING_DIR)/usr/local/ssl/fips-2.0/bin/incore
+endif
+
 ifeq ($(BR2_PACKAGE_OPENSSL_BIN),)
 define OPENSSL_DISABLE_APPS
 	$(SED) '/^build_apps/! s/build_apps//' $(@D)/Makefile.org
@@ -56,6 +65,7 @@
 	(cd $(@D); \
 		$(TARGET_CONFIGURE_ARGS) \
 		$(TARGET_CONFIGURE_OPTS) \
+		$(OPENSSL_FIPS_OPT) \
 		./Configure \
 			linux-$(OPENSSL_TARGET_ARCH) \
 			--prefix=/usr \
@@ -70,21 +80,22 @@
 			enable-tlsext \
 			$(if $(BR2_PREFER_STATIC_LIB),zlib,zlib-dynamic) \
 			$(if $(BR2_PREFER_STATIC_LIB),no-dso) \
+			$(OPENSSL_FIPS_CFG) \
 	)
 	$(SED) "s:-march=[-a-z0-9] ::" -e "s:-mcpu=[-a-z0-9] ::g" $(@D)/Makefile
 	$(SED) "s:-O[0-9]:$(OPENSSL_CFLAGS):" $(@D)/Makefile
 endef
 
 define OPENSSL_BUILD_CMDS
-	$(MAKE1) -C $(@D)
+	$(MAKE1) $(OPENSSL_FIPS_MAKE_OPT) -C $(@D)
 endef
 
 define OPENSSL_INSTALL_STAGING_CMDS
-	$(MAKE1) -C $(@D) INSTALL_PREFIX=$(STAGING_DIR) install
+	$(MAKE1) $(OPENSSL_FIPS_MAKE_OPT) -C $(@D) INSTALL_PREFIX=$(STAGING_DIR) install
 endef
 
 define OPENSSL_INSTALL_TARGET_CMDS
-	$(MAKE1) -C $(@D) INSTALL_PREFIX=$(TARGET_DIR) install
+	$(MAKE1) $(OPENSSL_FIPS_MAKE_OPT) -C $(@D) INSTALL_PREFIX=$(TARGET_DIR) install
 endef
 
 define OPENSSL_REMOVE_DEV_FILES
@@ -96,7 +107,6 @@
 endif
 
 define OPENSSL_INSTALL_FIXUPS
-	rm -f $(TARGET_DIR)/usr/bin/c_rehash
 endef

 OPENSSL_POST_INSTALL_TARGET_HOOKS += OPENSSL_INSTALL_FIXUPS
diff -Naur old/package/pppd/pppd-2.4.5-maxconfreq-30000.patch new/package/pppd/pppd-2.4.5-maxconfreq-30000.patch
--- old/package/pppd/pppd-2.4.5-maxconfreq-30000.patch	1970-01-01 01:00:00.000000000 +0100
+++ new/package/pppd/pppd-2.4.5-maxconfreq-30000.patch	2013-04-22 21:36:40.000000000 +0100
@@ -0,0 +1,12 @@
+diff -urN ppp-2.4.5/pppd/fsm.h ppp-2.4.5-lt1/pppd/fsm.h
+--- ppp-2.4.5/pppd/fsm.h	2009-11-16 17:26:07.000000000 -0500
++++ ppp-2.4.5-lt1/pppd/fsm.h	2012-09-04 22:39:37.405603914 -0400
+@@ -145,7 +145,7 @@
+  */
+ #define DEFTIMEOUT	3	/* Timeout time in seconds */
+ #define DEFMAXTERMREQS	2	/* Maximum Terminate-Request transmissions */
+-#define DEFMAXCONFREQS	10	/* Maximum Configure-Request transmissions */
++#define DEFMAXCONFREQS	30000	/* Maximum Configure-Request transmissions */
+ #define DEFMAXNAKLOOPS	5	/* Maximum number of nak loops */
+ 
+ 
diff -Naur old/package/proftpd/proftpd.conf new/package/proftpd/proftpd.conf
--- old/package/proftpd/proftpd.conf	1970-01-01 01:00:00.000000000 +0100
+++ new/package/proftpd/proftpd.conf	2013-04-22 21:36:40.000000000 +0100
@@ -0,0 +1,68 @@
+# This is a basic ProFTPD configuration file (rename it to 
+# 'proftpd.conf' for actual use.  It establishes a single server
+# and a single anonymous login.  It assumes that you have a user/group
+# "nobody" and "ftp" for normal operation and anon.
+
+ServerName			"ProFTPD Default Installation"
+ServerType			standalone
+DefaultServer			on
+UseReverseDNS			off
+IdentLookups			off
+ServerIdent on			"FTP Server Ready."
+
+# Port 21 is the standard FTP port.
+Port				21
+
+# Don't use IPv6 support by default.
+UseIPv6				off
+
+# Umask 022 is a good standard umask to prevent new dirs and files
+# from being group and world writable.
+Umask				022
+
+# To prevent DoS attacks, set the maximum number of child processes
+# to 30.  If you need to allow more than 30 concurrent connections
+# at once, simply increase this value.  Note that this ONLY works
+# in standalone mode, in inetd mode you should use an inetd server
+# that allows you to limit maximum number of processes per service
+# (such as xinetd).
+MaxInstances			30
+
+# Set the user and group under which the server will run.
+User				nobody
+Group				nogroup
+
+# To cause every FTP user to be "jailed" (chrooted) into their home
+# directory, uncomment this line.
+#DefaultRoot ~
+
+# Normally, we want files to be overwriteable.
+AllowOverwrite		on
+
+# Bar use of SITE CHMOD by default
+<Limit SITE_CHMOD>
+  DenyAll
+</Limit>
+
+# A basic anonymous configuration, no upload directories.  If you do not
+# want anonymous users, simply delete this entire <Anonymous> section.
+<Anonymous ~ftp>
+  User				ftp
+  Group				ftp
+
+  # We want clients to be able to login with "anonymous" as well as "ftp"
+  UserAlias			anonymous ftp
+
+  # Limit the maximum number of anonymous logins
+  MaxClients			10
+
+  # We want 'welcome.msg' displayed at login, and '.message' displayed
+  # in each newly chdired directory.
+  DisplayLogin			welcome.msg
+  DisplayChdir			.message
+
+  # Limit WRITE everywhere in the anonymous chroot
+  <Limit WRITE>
+    DenyAll
+  </Limit>
+</Anonymous>
diff -Naur old/package/proftpd/proftpd.mk new/package/proftpd/proftpd.mk
--- old/package/proftpd/proftpd.mk	2013-04-22 23:49:35.121440173 +0100
+++ new/package/proftpd/proftpd.mk	2013-04-22 21:36:40.000000000 +0100
@@ -34,12 +34,12 @@
 		$(INSTALL) -m 0644 -D $(@D)/sample-configurations/basic.conf $(TARGET_DIR)/etc/proftpd.conf; \
 		$(if $(BR2_INET_IPV6),,$(SED) 's/^UseIPv6/# UseIPv6/' $(TARGET_DIR)/etc/proftpd.conf;) \
 	fi
-	$(INSTALL) -m 0755 package/proftpd/S50proftpd $(TARGET_DIR)/etc/init.d
+	$(INSTALL) -m 0755 -D package/proftpd/S50proftpd $(TARGET_DIR)/etc/init.d/opt/S50proftpd
 endef
 
 define PROFTPD_UNINSTALL_TARGET_CMDS
 	rm -f $(TARGET_DIR)/$(PROFTPD_TARGET_BINARY)
-	rm -f $(TARGET_DIR)/etc/init.d/S50proftpd
+	rm -f $(TARGET_DIR)/etc/init.d/opt/S50proftpd
 	rm -f $(TARGET_DIR)/etc/proftpd.conf
 endef
 
diff -Naur old/package/proftpd/S50proftpd new/package/proftpd/S50proftpd
--- old/package/proftpd/S50proftpd	2013-04-22 23:49:34.941440166 +0100
+++ new/package/proftpd/S50proftpd	2013-05-09 10:52:16.741777628 -0500
@@ -1,46 +1,58 @@
 #!/bin/sh 
+# proftpd - ftp daemon init-script
+# supports inetd and pre-checking configuration
 
-DAEMON=/usr/sbin/proftpd
 trap "" 1
 trap "" 15
-test -f $DAEMON || exit 0
 [ ! -d /var/run/proftpd ] && mkdir /var/run/proftpd
 [ ! -f /var/log/wtmp ] && touch /var/log/wtmp
 
-start() {
-	echo -n "Starting ProFTPD: "
-	$DAEMON
-	if [ $? != 0 ]; then
-		echo "FAILED"
-		exit 1
-	else
-		echo "done"
-	fi
-}
-
-stop() {
-	echo -n "Stopping ProFTPD: "
-	killall proftpd
-        echo "done"
+pre_check() {
+  [ -s /etc/proftpd.conf ] \
+  || { on_error="config missing"; return 1; }
+  
+  if [ -z "$inetd_" ]
+  then
+    # set ServerType to "standalone" if necessary
+    sed '/^ServerType[\t ]*inetd/s/inetd/standalone/' -i /etc/proftpd.conf
+  else
+    # set ServerType to "inetd" if necessary
+    sed '/^ServerType[\t ]*standalone/s/standalone/inetd/' -i /etc/proftpd.conf
+  fi
 }
 
 case "$1" in
-    start)
-	start
-	;;
-
-    stop)
-	stop
-	;;
-
-    restart)
-    	stop
-    	start
-	;;
-
-    *)
-	echo "Usage: /etc/init.d/S50proftpd {start|stop|restart}"
-	exit 1
+  stop)
+    echo -n "Stopping ProFTPD: "
+    killall proftpd 2>/dev/null
+    echo "OK"
+    ;;
+
+  start)
+    echo -n "Starting ProFTPD: "
+    [ -z "$rcS_" ] && pre_check
+    /usr/sbin/proftpd
+    if [ $? != 0 ]
+    then
+      echo "FAILED ${on_error:+- $on_error}"
+      false
+    else
+      echo "OK"
+    fi
+    ;;
+
+  restart)
+    $0 stop
+    $0 start
+    ;;
+
+  check)
+    pre_check
+    ;;
+  
+  *)
+    echo "Usage: $0 {stop|start|restart}"
+    false
 	;;
 esac
 
diff -Naur old/package/udev/S10udev new/package/udev/S10udev
--- old/package/udev/S10udev	2013-04-22 23:49:35.129440172 +0100
+++ new/package/udev/S10udev	2013-04-22 21:36:40.000000000 +0100
@@ -1,48 +1,56 @@
 #!/bin/sh
 #
-# udev	This is a minimal non-LSB version of a UDEV startup script.  It
-#	was derived by stripping down the udev-058 LSB version for use
-#	with buildroot on embedded hardware using Linux 2.6.34+ kernels.
+#   This is a minimal non-LSB version of a UDEV startup script.  It
+#   was derived by stripping down the udev-058 LSB version for use
+#   with buildroot on embedded hardware using Linux 2.6.34+ kernels.
 #
-#	You may need to customize this for your system's resource limits
-#	(including startup time!) and administration.  For example, if
-#	your early userspace has a custom initramfs or initrd you might
-#	need /dev much earlier; or without hotpluggable busses (like USB,
-#	PCMCIA, MMC/SD, and so on) your /dev might be static after boot.
+#   You may need to customize this for your system's resource limits
+#   (including startup time!) and administration.  For example, if
+#   your early userspace has a custom initramfs or initrd you might
+#   need /dev much earlier; or without hotpluggable busses (like USB,
+#   PCMCIA, MMC/SD, and so on) your /dev might be static after boot.
 #
-#	This script assumes your system boots right into the eventual root
-#	filesystem, and that init runs this udev script before any programs
-#	needing more device nodes than the bare-bones set -- /dev/console,
-#	/dev/zero, /dev/null -- that's needed to boot and run this script.
+#   This script assumes your system boots right into the eventual root
+#   filesystem, and that init runs this udev script before any programs
+#   needing more device nodes than the bare-bones set -- /dev/console,
+#   /dev/zero, /dev/null -- that's needed to boot and run this script.
 #
 
-# Check for missing binaries
+# check for missing binaries
 UDEV_BIN=/lib/udev/udevd
 test -x $UDEV_BIN || exit 5
 
-# Check for config file and read it
-UDEV_CONFIG=/etc/udev/udev.conf
-test -r $UDEV_CONFIG || exit 6
-. $UDEV_CONFIG
-
-case "$1" in
-    start)
-        echo -n "Populating ${udev_root:-/dev} using udev: "
-        echo -e '\000\000\000\000' > /proc/sys/kernel/hotplug
-        $UDEV_BIN -d || (echo "FAIL" && exit 1)
-        udevadm trigger
-        echo "done"
-        ;;
-    stop)
-        # Stop execution of events
-        udevadm control --stop-exec-queue
-        killall udevd
-        ;;
-    *)
-        echo "Usage: $0 {start|stop}"
-        exit 1
-        ;;
-esac
+# check for config file and read it
+UDEV_CONF=/etc/udev/udev.conf
+test -r $UDEV_CONF || exit 6
+source $UDEV_CONF
+
+case $1 in
+
+  stop)
+    echo -n "Stopping execution of events, via udev... "
+    udevadm control --stop-exec-queue
+    killall udevd 2>/dev/null
+    echo
+    ;;
 
+  start)
+    echo -n "Starting udev daemon, populating ${udev_root:-/dev} "
+    echo -e '\000\000\000\000' >/proc/sys/kernel/hotplug
+    $UDEV_BIN -d || { echo "ERROR"; exit 1; }
+    udevadm trigger
+    echo
+    ;;
+
+  *)
+    echo "Usage: $0 {stop|start}"
+    exit 1
+    ;;
+esac
 
 exit 0
+##
+## report stuff relative to a device node
+##  udevadm info -a -n </dev/device>
+##
+
diff -Naur old/package/udev/udev.mk new/package/udev/udev.mk
--- old/package/udev/udev.mk	2013-04-22 23:49:35.129440172 +0100
+++ new/package/udev/udev.mk	2013-04-22 21:36:40.000000000 +0100
@@ -43,6 +43,18 @@
 	$(INSTALL) -m 0755 package/udev/S10udev $(TARGET_DIR)/etc/init.d/S10udev
 endef
 
+define UDEV_NO_AUTO_LOAD_HANDLING
+	@echo UDEV_NO_AUTO_LOAD_HANDLING 
+	# Avoid auto-loading all device drivers, with udev_182. (HACK)
+	# Some devices require firmware and are highly dependent on various conditions
+	# and thus handled by specific scripts.
+	( cd $(TARGET_DIR)/lib/udev; \
+	  [ -f rules.d/??-drivers.rules ] \
+	  && { mkdir -p rules.disabled; mv rules.d/??-drivers.rules rules.disabled; } || : \
+	)
+endef
+
 UDEV_POST_INSTALL_TARGET_HOOKS += UDEV_INSTALL_INITSCRIPT
+UDEV_POST_INSTALL_TARGET_HOOKS += UDEV_NO_AUTO_LOAD_HANDLING
 
 $(eval $(autotools-package))
diff -Naur old/package/usbmount/usbmount.mk new/package/usbmount/usbmount.mk
--- old/package/usbmount/usbmount.mk	2013-04-22 23:49:35.125440173 +0100
+++ new/package/usbmount/usbmount.mk	2013-04-22 21:36:40.000000000 +0100
@@ -16,7 +16,7 @@
 	$(INSTALL) -m 0755 -D $(@D)/00_remove_model_symlink 	\
 		$(TARGET_DIR)/etc/usbmount/usbmount.d/00_remove_model_symlink
 
-	$(INSTALL) -m 0644 -D $(@D)/usbmount.rules $(TARGET_DIR)/lib/udev/rules.d/usbmount.rules
+	$(INSTALL) -m 0644 -D $(@D)/usbmount.rules $(TARGET_DIR)/etc/udev/rules.d/usbmount.rules
 	@if [ ! -f $(TARGET_DIR)/etc/usbmount/usbmount.conf ]; then \
 	        $(INSTALL) -m 0644 -D $(@D)/usbmount.conf $(TARGET_DIR)/etc/usbmount/usbmount.conf; \
 	fi
@@ -27,7 +27,7 @@
 define USBMOUNT_UNINSTALL_TARGET_CMDS
 	rm -rf $(TARGET_DIR)/etc/usbmount			\
 		$(TARGET_DIR)/usr/share/usbmount/usbmount	\
-		$(TARGET_DIR)/lib/udev/rules.d/usbmount.rules	\
+		$(TARGET_DIR)/etc/udev/rules.d/usbmount.rules	\
 		$(TARGET_DIR)/media/usb?
 endef
 
diff -Naur old/system/device_table_dev.txt new/system/device_table_dev.txt
--- old/system/device_table_dev.txt	2013-04-22 23:49:34.925440166 +0100
+++ new/system/device_table_dev.txt	2013-04-22 21:36:40.000000000 +0100
@@ -46,8 +46,8 @@

 
 # MTD stuff
-/dev/mtd	c	640	0	0	90	0	0	2	4
-/dev/mtdblock	b	640	0	0	31	0	0	1	4
+/dev/mtd	c	640	0	0	90	0	0	2	8
+/dev/mtdblock	b	640	0	0	31	0	0	1	8
 
 #Tun/tap driver
 /dev/net	d	755	0	0	-	-	-	-	-
