--- buildroot/boot/uboot/Config.in.old	2012-12-04 09:11:09.357921576 +0000
+++ buildroot/boot/uboot/Config.in	2012-12-04 09:15:31.373931301 +0000
@@ -16,6 +16,9 @@
 	help
 	  Select the specific U-Boot version you want to use
 
+config BR2_TARGET_UBOOT_2011_09
+	bool "2011.09"
+
 config BR2_TARGET_UBOOT_2011_06
 	bool "2011.06"
 
@@ -48,6 +51,7 @@
 
 config BR2_TARGET_UBOOT_VERSION
 	string
+	default "2011.09"	if BR2_TARGET_UBOOT_2011_09
 	default "2011.06"	if BR2_TARGET_UBOOT_2011_06
 	default "2011.03"	if BR2_TARGET_UBOOT_2011_03
 	default "2010.12"	if BR2_TARGET_UBOOT_2010_12
